// plus-pro-components/es/components/date-picker/src/index.vue2.mjs
import { defineComponent, ref, reactive, watch, withDirectives, openBlock, createElementBlock, normalizeClass, createVNode, unref, mergeProps, createElementVNode, toDisplayString } from 'vue';
import { ElDatePicker, ClickOutside } from 'element-plus';
import '../../utils/index.mjs';
import '../../../hooks/index.mjs';
import { useLocale } from '../../../hooks/useLocale.mjs';
import { isFunction } from '../../utils/is.mjs';

const _hoisted_1 = { class: "plus-date-picker__middle" };
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "PlusDatePicker"
  },
  __name: "index",
  props: {
    modelValue: { type: Array, required: false, default: () => [] },
    format: { type: String, required: false, default: "YYYY-MM-DD HH:mm:ss" },
    valueFormat: { type: String, required: false, default: "YYYY-MM-DD HH:mm:ss" },
    rangeSeparator: { type: String, required: false, default: "/" },
    type: { type: String, required: false, default: "datetime" },
    startProps: { type: null, required: false, default: () => ({}) },
    endProps: { type: null, required: false, default: () => ({}) },
    startDisabledDate: { type: Function, required: false, default: (startTime, endValue) => {
      if (!endValue)
        return false;
      return startTime.getTime() > new Date(endValue).getTime();
    } },
    endDisabledDate: { type: Function, required: false, default: (endTime, startValue) => {
      if (!startValue)
        return false;
      return endTime.getTime() < new Date(startValue).getTime();
    } }
  },
  emits: ["change", "focus", "update:modelValue"],
  setup(__props, { expose: __expose, emit }) {
    const props = __props;
    const { t } = useLocale();
    const startPickerInstance = ref();
    const endPickerInstance = ref();
    const state = reactive({
      start: "",
      end: ""
    });
    const isFocus = ref(false);
    const handleFocus = (event) => {
      isFocus.value = true;
      emit("focus", event);
    };
    const onClickOutside = () => {
      isFocus.value = false;
    };
    const subStartDisabledDate = (time) => {
      if (props.startDisabledDate && isFunction(props.startDisabledDate)) {
        return props.startDisabledDate(time, state.end);
      }
      return false;
    };
    const subEndDisabledDate = (time) => {
      if (props.endDisabledDate && isFunction(props.endDisabledDate)) {
        return props.endDisabledDate(time, state.start);
      }
      return false;
    };
    watch(
      () => props.modelValue,
      (val) => {
        const [start, end] = val;
        state.start = start;
        state.end = end;
      },
      {
        immediate: true
      }
    );
    const handleChange = () => {
      const res = [state.start, state.end];
      emit("update:modelValue", res);
      emit("change", res);
    };
    __expose({
      startPickerInstance,
      endPickerInstance
    });
    return (_ctx, _cache) => {
      return withDirectives((openBlock(), createElementBlock(
        "div",
        {
          class: normalizeClass(["plus-date-picker", {
            "is-focus": isFocus.value
          }])
        },
        [
          createVNode(unref(ElDatePicker), mergeProps({
            ref_key: "startPickerInstance",
            ref: startPickerInstance,
            modelValue: state.start,
            "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => state.start = $event),
            type: "datetime",
            placeholder: unref(t)("plus.datepicker.startPlaceholder"),
            format: _ctx.format,
            "value-format": _ctx.valueFormat,
            "disabled-date": subStartDisabledDate,
            class: "plus-date-picker__start",
            clearable: ""
          }, _ctx.startProps, {
            onChange: handleChange,
            onFocus: handleFocus
          }), null, 16, ["modelValue", "placeholder", "format", "value-format"]),
          createElementVNode(
            "span",
            _hoisted_1,
            toDisplayString(_ctx.rangeSeparator),
            1
            /* TEXT */
          ),
          createVNode(unref(ElDatePicker), mergeProps({
            ref_key: "endPickerInstance",
            ref: endPickerInstance,
            modelValue: state.end,
            "onUpdate:modelValue": _cache[1] || (_cache[1] = ($event) => state.end = $event),
            type: "datetime",
            format: _ctx.format,
            "value-format": _ctx.valueFormat,
            placeholder: unref(t)("plus.datepicker.endPlaceholder"),
            "disabled-date": subEndDisabledDate,
            class: "plus-date-picker__end",
            clearable: ""
          }, _ctx.endProps, {
            onChange: handleChange,
            onFocus: handleFocus
          }), null, 16, ["modelValue", "format", "value-format", "placeholder"])
        ],
        2
        /* CLASS */
      )), [
        [unref(ClickOutside), onClickOutside]
      ]);
    };
  }
});

export { _sfc_main as default };
