// plus-pro-components/es/components/layout/src/index.vue2.mjs
import { defineComponent, ref, computed, openBlock, createBlock, unref, withCtx, createCommentVNode, createVNode, normalizeProps, guardReactiveProps, createSlots, renderSlot, mergeProps } from 'vue';
import { PlusBreadcrumb } from '../../breadcrumb/index.mjs';
import { PlusSidebar } from '../../sidebar/index.mjs';
import { PlusHeader } from '../../header/index.mjs';
import '../../utils/index.mjs';
import { ElContainer, ElMain, ElScrollbar, ElBacktop } from 'element-plus';
import { isPlainObject } from '../../utils/is.mjs';

var _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "PlusLayout"
  },
  __name: "index",
  props: {
    sidebarProps: { type: null, required: false, default: void 0 },
    headerProps: { type: null, required: false, default: void 0 },
    breadcrumbProps: { type: null, required: false, default: void 0 },
    scrollbarProps: { type: Object, required: false, default: () => ({
      always: true
    }) },
    backtop: { type: [Boolean, Object], required: false, default: true }
  },
  setup(__props, { expose: __expose }) {
    const props = __props;
    const PlusBreadcrumb$1 = PlusBreadcrumb;
    const PlusSidebar$1 = PlusSidebar;
    const PlusHeader$1 = PlusHeader;
    const plusSidebarInstance = ref();
    const backtopProps = computed(
      () => isPlainObject(props.backtop) ? props.backtop : {}
    );
    __expose({
      plusSidebarInstance
    });
    return (_ctx, _cache) => {
      return openBlock(), createBlock(unref(ElContainer), { class: "plus-layout" }, {
        default: withCtx(() => [
          createCommentVNode(" \u5934\u90E8 "),
          createVNode(
            unref(PlusHeader$1),
            normalizeProps(guardReactiveProps(_ctx.headerProps)),
            createSlots({
              _: 2
              /* DYNAMIC */
            }, [
              _ctx.$slots["header-left"] ? {
                name: "header-left",
                fn: withCtx((data) => [
                  renderSlot(_ctx.$slots, "header-left", normalizeProps(guardReactiveProps(data)))
                ]),
                key: "0"
              } : void 0,
              _ctx.$slots["header-right"] ? {
                name: "header-right",
                fn: withCtx((data) => [
                  renderSlot(_ctx.$slots, "header-right", normalizeProps(guardReactiveProps(data)))
                ]),
                key: "1"
              } : void 0
            ]),
            1040
            /* FULL_PROPS, DYNAMIC_SLOTS */
          ),
          createVNode(unref(ElContainer), null, {
            default: withCtx(() => [
              createCommentVNode(" \u4FA7\u8FB9\u680F "),
              createVNode(
                unref(PlusSidebar$1),
                mergeProps(_ctx.sidebarProps, {
                  ref_key: "plusSidebarInstance",
                  ref: plusSidebarInstance
                }),
                createSlots({
                  _: 2
                  /* DYNAMIC */
                }, [
                  _ctx.$slots["sidebar-extra"] ? {
                    name: "sidebar-extra",
                    fn: withCtx((data) => [
                      renderSlot(_ctx.$slots, "sidebar-extra", normalizeProps(guardReactiveProps(data)))
                    ]),
                    key: "0"
                  } : void 0,
                  _ctx.$slots["sidebar-item"] ? {
                    name: "sidebar-item",
                    fn: withCtx((data) => [
                      renderSlot(_ctx.$slots, "sidebar-item", normalizeProps(guardReactiveProps(data)))
                    ]),
                    key: "1"
                  } : void 0,
                  _ctx.$slots["sidebar-sub"] ? {
                    name: "sidebar-sub",
                    fn: withCtx((data) => [
                      renderSlot(_ctx.$slots, "sidebar-sub", normalizeProps(guardReactiveProps(data)))
                    ]),
                    key: "2"
                  } : void 0,
                  _ctx.$slots["sidebar-item-title"] ? {
                    name: "sidebar-item-title",
                    fn: withCtx((data) => [
                      renderSlot(_ctx.$slots, "sidebar-item-title", normalizeProps(guardReactiveProps(data)))
                    ]),
                    key: "3"
                  } : void 0
                ]),
                1040
                /* FULL_PROPS, DYNAMIC_SLOTS */
              ),
              createCommentVNode(" \u4E3B\u5185\u5BB9 "),
              createVNode(unref(ElMain), null, {
                default: withCtx(() => [
                  createVNode(
                    unref(ElScrollbar),
                    mergeProps({ class: "plus-layout-main__scrollbar" }, _ctx.scrollbarProps),
                    {
                      default: withCtx(() => [
                        createCommentVNode(" \u9762\u5305\u5C51\u4E0A\u65B9 "),
                        _ctx.$slots["layout-extra"] ? renderSlot(_ctx.$slots, "layout-extra", { key: 0 }) : createCommentVNode("v-if", true),
                        createCommentVNode(" \u9762\u5305\u5C51 "),
                        createVNode(
                          unref(PlusBreadcrumb$1),
                          normalizeProps(guardReactiveProps(_ctx.breadcrumbProps)),
                          createSlots({
                            _: 2
                            /* DYNAMIC */
                          }, [
                            _ctx.$slots["breadcrumb-item-title"] ? {
                              name: "breadcrumb-item-title",
                              fn: withCtx((data) => [
                                renderSlot(_ctx.$slots, "breadcrumb-item-title", normalizeProps(guardReactiveProps(data)))
                              ]),
                              key: "0"
                            } : void 0
                          ]),
                          1040
                          /* FULL_PROPS, DYNAMIC_SLOTS */
                        ),
                        renderSlot(_ctx.$slots, "default"),
                        _ctx.backtop ? (openBlock(), createBlock(
                          unref(ElBacktop),
                          mergeProps({ key: 1 }, backtopProps.value, { target: ".plus-layout .plus-layout-main__scrollbar" }),
                          null,
                          16
                          /* FULL_PROPS */
                        )) : createCommentVNode("v-if", true)
                      ]),
                      _: 3
                      /* FORWARDED */
                    },
                    16
                    /* FULL_PROPS */
                  )
                ]),
                _: 3
                /* FORWARDED */
              })
            ]),
            _: 3
            /* FORWARDED */
          })
        ]),
        _: 3
        /* FORWARDED */
      });
    };
  }
});

export { _sfc_main as default };
