// plus-pro-components/es/index.mjs
import installer from './defaults.mjs';
import './components/index.mjs';
import './hooks/index.mjs';
import './types/index.mjs';
import './constants/index.mjs';
export { PlusDialog } from './components/dialog/index.mjs';
export { PlusPagination } from './components/pagination/index.mjs';
export { PlusTable } from './components/table/index.mjs';
export { PlusRadio } from './components/radio/index.mjs';
export { PlusDatePicker } from './components/date-picker/index.mjs';
export { PlusDescriptions } from './components/descriptions/index.mjs';
export { PlusDisplayItem } from './components/display-item/index.mjs';
export { PlusForm } from './components/form/index.mjs';
export { PlusFormItem } from './components/form-item/index.mjs';
export { PlusSearch } from './components/search/index.mjs';
export { PlusDialogForm } from './components/dialog-form/index.mjs';
export { PlusDrawerForm } from './components/drawer-form/index.mjs';
export { PlusPage } from './components/page/index.mjs';
export { PlusPopover } from './components/popover/index.mjs';
export { PlusStepsForm } from './components/steps-form/index.mjs';
export { PlusInputTag } from './components/input-tag/index.mjs';
export { PlusBreadcrumb } from './components/breadcrumb/index.mjs';
export { PlusSidebar, PlusSidebarItem } from './components/sidebar/index.mjs';
export { PlusHeader } from './components/header/index.mjs';
export { PlusLayout } from './components/layout/index.mjs';
export { PlusCheckCard } from './components/check-card/index.mjs';
export { PlusCheckCardGroup } from './components/check-card-group/index.mjs';
export { useGetOptions } from './hooks/useGetOptions.mjs';
export { buildLocaleContext, buildTranslator, translate, useLocale } from './hooks/useLocale.mjs';
export { useTable } from './hooks/useTable.mjs';
export { DefaultPageInfo, DefaultPageSizeList } from './constants/page.mjs';
export { TableFormRefInjectionKey } from './constants/form.mjs';

const install = installer.install;

export { installer as default, install };
