// v-contextmenu@3.2.0/es/components/ContextmenuItem.jsx
import { computed, defineComponent, inject, ref } from 'vue';
import { CLASSES } from '../constants';
var ContextmenuItem = defineComponent({
    name: 'VContextmenuItem',
    props: {
        disabled: {
            type: Boolean,
            default: false,
        },
        hideOnClick: {
            type: Boolean,
            default: true,
        },
        contextmenuAsClick: {
            type: Boolean,
            default: true,
        },
    },
    emits: ['click', 'contextmenu', 'mouseenter', 'mouseleave'],
    setup: function (props, _a) {
        var emit = _a.emit;
        var rootHide = inject('hide');
        var hover = ref(false);
        var classes = computed(function () {
            var _a;
            return (_a = {},
                _a[CLASSES.contextmenuItem] = true,
                _a[CLASSES.contextmenuItemDisabled] = props.disabled,
                _a[CLASSES.contextmenuItemHover] = hover.value,
                _a);
        });
        var handleClick = function (evt) {
            emit('click', evt);
            if (props.disabled)
                return;
            props.hideOnClick && (rootHide === null || rootHide === void 0 ? void 0 : rootHide());
        };
        var handleContextmenu = function (evt) {
            emit('contextmenu', evt);
            if (props.contextmenuAsClick) {
                if (props.disabled)
                    return;
                props.hideOnClick && (rootHide === null || rootHide === void 0 ? void 0 : rootHide());
            }
        };
        var handleMouseenter = function (evt) {
            emit('mouseenter', evt);
            if (props.disabled)
                return;
            hover.value = true;
        };
        var handleMouseleave = function (evt) {
            emit('mouseleave', evt);
            if (props.disabled)
                return;
            hover.value = false;
        };
        return {
            classes: classes,
            handleClick: handleClick,
            handleContextmenu: handleContextmenu,
            handleMouseenter: handleMouseenter,
            handleMouseleave: handleMouseleave,
        };
    },
    render: function () {
        var _a, _b;
        return (<li class={this.classes} onClick={this.handleClick} onContextmenu={this.handleContextmenu} onMouseenter={this.handleMouseenter} onMouseleave={this.handleMouseleave}>
        {(_b = (_a = this.$slots).default) === null || _b === void 0 ? void 0 : _b.call(_a)}
      </li>);
    },
});
export default ContextmenuItem;
