// plus-pro-components/es/components/steps-form/src/index.vue2.mjs
import { defineComponent, ref, openBlock, createElementBlock, normalizeClass, createVNode, unref, mergeProps, withCtx, Fragment, renderList, createBlock, createSlots, renderSlot } from 'vue';
import { cloneDeep } from 'lodash-es';
import { ElSteps, ElStep } from 'element-plus';
import '../../../hooks/index.mjs';
import { PlusForm } from '../../form/index.mjs';
import { useLocale } from '../../../hooks/useLocale.mjs';

var _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "PlusStepsForm"
  },
  __name: "index",
  props: {
    modelValue: { type: Number, required: true, default: 1 },
    data: { type: Array, required: true, default: () => [] }
  },
  emits: ["pre", "next", "update:modelValue", "change"],
  setup(__props, { emit }) {
    const props = __props;
    const { t } = useLocale();
    const active = ref(cloneDeep(props.modelValue));
    const handleChange = (values, column) => {
      emit("change", values, column);
    };
    const pre = () => {
      if (active.value-- > props.data.length + 1)
        active.value = 1;
      emit("pre", active.value);
      emit("update:modelValue", active.value);
    };
    const next = (values) => {
      if (active.value++ > props.data.length - 1)
        active.value = props.data.length;
      emit("next", active.value, values);
      emit("update:modelValue", active.value);
    };
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "div",
        {
          class: normalizeClass(["plus-steps-form", _ctx.$attrs.direction === "vertical" ? "plus-steps-from-vertical" : ""])
        },
        [
          createVNode(unref(ElSteps), mergeProps({
            active: active.value,
            "finish-status": "success"
          }, _ctx.$attrs), {
            default: withCtx(() => [
              (openBlock(true), createElementBlock(
                Fragment,
                null,
                renderList(_ctx.data, (item) => {
                  return openBlock(), createBlock(
                    unref(ElStep),
                    mergeProps({
                      key: item.title
                    }, item),
                    createSlots({
                      _: 2
                      /* DYNAMIC */
                    }, [
                      _ctx.$slots.icon ? {
                        name: "icon",
                        fn: withCtx(() => [
                          renderSlot(_ctx.$slots, "icon", {
                            icon: item.icon,
                            title: item.title,
                            description: item.description
                          })
                        ]),
                        key: "0"
                      } : void 0,
                      _ctx.$slots.title ? {
                        name: "title",
                        fn: withCtx(() => [
                          renderSlot(_ctx.$slots, "title", {
                            icon: item.icon,
                            title: item.title,
                            description: item.description
                          })
                        ]),
                        key: "1"
                      } : void 0,
                      _ctx.$slots.description ? {
                        name: "description",
                        fn: withCtx(() => [
                          renderSlot(_ctx.$slots, "description", {
                            icon: item.icon,
                            title: item.title,
                            description: item.description
                          })
                        ]),
                        key: "2"
                      } : void 0
                    ]),
                    1040
                    /* FULL_PROPS, DYNAMIC_SLOTS */
                  );
                }),
                128
                /* KEYED_FRAGMENT */
              ))
            ]),
            _: 3
            /* FORWARDED */
          }, 16, ["active"]),
          createVNode(unref(PlusForm), mergeProps(_ctx.data[active.value - 1].form, {
            "has-reset": active.value !== 1,
            "submit-text": active.value === _ctx.data.length ? unref(t)("plus.stepsForm.submitText") : unref(t)("plus.stepsForm.nextText"),
            "reset-text": unref(t)("plus.stepsForm.preText"),
            onSubmit: next,
            onReset: pre,
            onChange: handleChange
          }), null, 16, ["has-reset", "submit-text", "reset-text"])
        ],
        2
        /* CLASS */
      );
    };
  }
});

export { _sfc_main as default };
