// v-contextmenu/es/components/ContextmenuGroup.jsx
import { computed, defineComponent } from 'vue';
import { CLASSES } from '../constants';
var ContextmenuGroup = defineComponent({
    name: 'VContextmenuGroup',
    props: {
        title: {
            type: String,
            default: undefined,
        },
        maxWidth: {
            type: [Number, String],
            default: undefined,
        },
    },
    setup: function (props) {
        var style = computed(function () {
            if (!props.maxWidth)
                return;
            return {
                'max-width': typeof props.maxWidth === 'number'
                    ? "".concat(props.maxWidth, "px")
                    : props.maxWidth,
                'overflow-x': 'auto',
            };
        });
        return {
            style: style,
        };
    },
    methods: {
        renderTitle: function () {
            var _a, _b;
            var content = ((_b = (_a = this.$slots).title) === null || _b === void 0 ? void 0 : _b.call(_a)) || this.title;
            return content ? (<div class={CLASSES.contextmenuGroupTitle}>{content}</div>) : null;
        },
    },
    render: function () {
        var _a, _b;
        return (<li class={CLASSES.contextmenuGroup}>
        {this.renderTitle()}

        <ul style={this.style} class={CLASSES.contextmenuGroupMenus}>
          {(_b = (_a = this.$slots).default) === null || _b === void 0 ? void 0 : _b.call(_a)}
        </ul>
      </li>);
    },
});
export default ContextmenuGroup;
