// plus-pro-components/es/components/display-item/src/index.vue2.mjs
import { defineComponent, ref, computed, watch, openBlock, createElementBlock, Fragment, createCommentVNode, createBlock, unref, mergeProps, createSlots, withCtx, renderSlot, createVNode, createElementVNode, renderList, resolveDynamicComponent, normalizeProps, guardReactiveProps, createTextVNode, toDisplayString, normalizeClass, normalizeStyle } from 'vue';
import { DocumentCopy, Select } from '@element-plus/icons-vue';
import { PlusForm } from '../../form/index.mjs';
import { getValue, setValue, getCustomProps, getFieldSlotName, getExtraSlotName, getTableCellSlotName } from '../../utils/index.mjs';
import '../../../hooks/index.mjs';
import { PlusRender } from '../../render/index.mjs';
import { ElImage, ElLink, ElTag, ElProgress, ElIcon } from 'element-plus';
import { useGetOptions } from '../../../hooks/useGetOptions.mjs';
import { isArray, isFunction } from '../../utils/is.mjs';
import { formatDate, formatMoney } from '../../utils/format.mjs';

const _hoisted_1 = ["innerHTML"];
const _hoisted_2 = { class: "plus-display-item" };
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "PlusDisplayItem"
  },
  __name: "index",
  props: {
    column: { type: null, required: true, default: () => ({ prop: "", label: "" }) },
    row: { type: null, required: true, default: () => ({}) },
    index: { type: Number, required: false, default: 0 }
  },
  emits: ["change"],
  setup(__props, { expose: __expose, emit }) {
    const props = __props;
    const isCellEdit = ref(false);
    const isForm = computed(() => props.column.editable === true || isCellEdit.value === true);
    const customFieldProps = ref({});
    const formInstance = ref();
    const options = useGetOptions(props.column);
    const columns = ref([]);
    const subRow = ref(props.row);
    const displayValue = computed({
      get() {
        return getValue(subRow.value, props.column.prop);
      },
      set(value) {
        setValue(subRow.value, props.column.prop, value);
      }
    });
    const modelValues = computed({
      get() {
        return { [props.column.prop]: displayValue.value };
      },
      set(values) {
        displayValue.value = values[props.column.prop];
      }
    });
    const params = computed(() => ({
      row: subRow.value,
      column: props.column,
      index: props.index
    }));
    const imageUrl = computed(() => {
      const option = displayValue.value;
      if (option && typeof option === "string") {
        return { options: [option], url: option };
      }
      if (isArray(option)) {
        return { options: option, url: option[0] };
      }
      return { options: [], url: "" };
    });
    const getStatus = computed(() => {
      var _a, _b, _c, _d;
      let option = (_a = options.value) == null ? void 0 : _a.find((i) => i.value === displayValue.value);
      if (((_b = props.column) == null ? void 0 : _b.customGetStatus) && isFunction((_c = props.column) == null ? void 0 : _c.customGetStatus)) {
        option = (_d = props.column) == null ? void 0 : _d.customGetStatus({
          options: options.value,
          value: displayValue.value,
          row: subRow.value
        });
      }
      if (!option) {
        return { label: "", value: "" };
      }
      return option;
    });
    watch(
      () => props.column,
      (val) => {
        if (val) {
          columns.value = [val];
        }
      },
      {
        immediate: true,
        deep: true
      }
    );
    watch(
      () => props.column.fieldProps,
      (val) => {
        getCustomProps(val, displayValue.value, subRow.value, props.index, "fieldProps").then((data) => {
          customFieldProps.value = data;
        }).catch((err) => {
          throw err;
        });
      },
      {
        immediate: true,
        deep: true
      }
    );
    watch(
      () => props.row,
      (val) => {
        subRow.value = { ...val };
      },
      {
        deep: true
      }
    );
    const copy = (data) => {
      const url = data;
      const textarea = document.createElement("textarea");
      textarea.readOnly = true;
      textarea.style.position = "absolute";
      textarea.style.left = "-9999px";
      textarea.value = url;
      document.body.appendChild(textarea);
      textarea.select();
      document.execCommand("Copy");
      textarea.remove();
    };
    const handelClickCopy = (item, row) => {
      copy(row[item.prop]);
      row.isCopy = true;
      setTimeout(() => {
        row.isCopy = false;
      }, 3e3);
    };
    const handleChange = (values) => {
      emit("change", { value: values[props.column.prop], prop: props.column.prop, row: subRow });
    };
    const startCellEdit = () => {
      isCellEdit.value = true;
    };
    const stopCellEdit = () => {
      isCellEdit.value = false;
    };
    const getDisplayItemInstance = () => {
      return {
        index: props.index,
        prop: props.column.prop,
        formInstance: computed(() => {
          var _a;
          return (_a = formInstance.value) == null ? void 0 : _a.formInstance;
        })
      };
    };
    __expose({
      startCellEdit,
      stopCellEdit,
      getDisplayItemInstance
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        Fragment,
        null,
        [
          createCommentVNode(" \u8868\u5355\u7B2C\u4E00\u4F18\u5148\u7EA7 "),
          isForm.value ? (openBlock(), createBlock(unref(PlusForm), mergeProps({
            key: 0,
            ref_key: "formInstance",
            ref: formInstance,
            modelValue: modelValues.value,
            "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => modelValues.value = $event),
            model: modelValues.value,
            columns: columns.value,
            "has-footer": false,
            "has-label": false
          }, _ctx.column.formProps, {
            class: "plus-display-item__form",
            onChange: handleChange
          }), createSlots({
            _: 2
            /* DYNAMIC */
          }, [
            _ctx.$slots[unref(getFieldSlotName)(_ctx.column.prop)] ? {
              name: unref(getFieldSlotName)(_ctx.column.prop),
              fn: withCtx((data) => [
                renderSlot(_ctx.$slots, unref(getFieldSlotName)(_ctx.column.prop), mergeProps(data, { row: subRow.value }))
              ]),
              key: "0"
            } : void 0,
            _ctx.$slots[unref(getExtraSlotName)(_ctx.column.prop)] ? {
              name: unref(getExtraSlotName)(_ctx.column.prop),
              fn: withCtx((data) => [
                renderSlot(_ctx.$slots, unref(getExtraSlotName)(_ctx.column.prop), mergeProps(data, { row: subRow.value }))
              ]),
              key: "1"
            } : void 0
          ]), 1040, ["modelValue", "model", "columns"])) : _ctx.column.render && unref(isFunction)(_ctx.column.render) ? (openBlock(), createElementBlock(
            Fragment,
            { key: 1 },
            [
              createCommentVNode(" \u81EA\u5B9A\u4E49\u663E\u793A "),
              createVNode(unref(PlusRender), {
                render: _ctx.column.render,
                params: params.value,
                "callback-value": displayValue.value,
                "custom-field-props": customFieldProps.value
              }, null, 8, ["render", "params", "callback-value", "custom-field-props"])
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.$slots[unref(getTableCellSlotName)(_ctx.column.prop)] ? (openBlock(), createElementBlock(
            Fragment,
            { key: 2 },
            [
              createCommentVNode(" \u63D2\u69FD "),
              renderSlot(_ctx.$slots, unref(getTableCellSlotName)(_ctx.column.prop), {
                prop: _ctx.column.prop,
                valueType: _ctx.column.valueType,
                row: subRow.value,
                value: displayValue.value,
                fieldProps: customFieldProps.value,
                column: _ctx.column
              })
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.renderHTML && unref(isFunction)(_ctx.column.renderHTML) ? (openBlock(), createElementBlock(
            Fragment,
            { key: 3 },
            [
              createCommentVNode("\u663E\u793AHTML "),
              createElementVNode("span", {
                class: "plus-display-item",
                innerHTML: _ctx.column.renderHTML(displayValue.value, { row: subRow.value, column: _ctx.column, index: _ctx.index })
              }, null, 8, _hoisted_1)
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "img" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 4 },
            [
              createCommentVNode("\u663E\u793A\u56FE\u7247 "),
              createVNode(unref(ElImage), mergeProps({
                class: "plus-display-item plus-display-item__image",
                fit: "cover",
                "preview-teleported": "",
                src: imageUrl.value.url,
                "preview-src-list": _ctx.column.preview !== false ? imageUrl.value.options : []
              }, customFieldProps.value), createSlots({
                _: 2
                /* DYNAMIC */
              }, [
                renderList(_ctx.column.fieldSlots, (fieldSlot, key) => {
                  return {
                    name: key,
                    fn: withCtx((data) => [
                      (openBlock(), createBlock(
                        resolveDynamicComponent(fieldSlot),
                        normalizeProps(guardReactiveProps(data)),
                        null,
                        16
                        /* FULL_PROPS */
                      ))
                    ])
                  };
                })
              ]), 1040, ["src", "preview-src-list"])
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "link" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 5 },
            [
              createCommentVNode("\u663E\u793A\u94FE\u63A5 "),
              createVNode(
                unref(ElLink),
                mergeProps({
                  type: "primary",
                  class: "plus-display-item plus-display-item__link"
                }, customFieldProps.value),
                createSlots({
                  default: withCtx(() => [
                    createTextVNode(
                      " " + toDisplayString(_ctx.column.linkText || displayValue.value),
                      1
                      /* TEXT */
                    )
                  ]),
                  _: 2
                  /* DYNAMIC */
                }, [
                  renderList(_ctx.column.fieldSlots, (fieldSlot, key) => {
                    return {
                      name: key,
                      fn: withCtx((data) => [
                        (openBlock(), createBlock(
                          resolveDynamicComponent(fieldSlot),
                          normalizeProps(guardReactiveProps(data)),
                          null,
                          16
                          /* FULL_PROPS */
                        ))
                      ])
                    };
                  })
                ]),
                1040
                /* FULL_PROPS, DYNAMIC_SLOTS */
              )
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "date-picker" && displayValue.value ? (openBlock(), createElementBlock(
            Fragment,
            { key: 6 },
            [
              createCommentVNode(" \u683C\u5F0F\u5316\u65F6\u95F4 "),
              createElementVNode(
                "span",
                mergeProps({ class: "plus-display-item" }, customFieldProps.value),
                toDisplayString(unref(formatDate)(displayValue.value)),
                17
                /* TEXT, FULL_PROPS */
              )
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "money" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 7 },
            [
              createCommentVNode(" \u683C\u5F0F\u5316\u91D1\u94B1 "),
              createElementVNode(
                "span",
                mergeProps({ class: "plus-display-item" }, customFieldProps.value),
                toDisplayString(unref(formatMoney)(displayValue.value)),
                17
                /* TEXT, FULL_PROPS */
              )
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "select" || _ctx.column.valueType === "radio" || _ctx.column.valueType === "checkbox" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 8 },
            [
              createCommentVNode(" \u72B6\u6001\u663E\u793A "),
              createElementVNode(
                "span",
                mergeProps({ class: "plus-display-item plus-display-item__badge" }, customFieldProps.value),
                [
                  getStatus.value.color || getStatus.value.type ? (openBlock(), createElementBlock(
                    "span",
                    {
                      key: 0,
                      class: normalizeClass([
                        "plus-display-item__badge__dot",
                        getStatus.value.type && !getStatus.value.color ? "plus-display-item__badge__dot--" + getStatus.value.type : ""
                      ]),
                      style: normalizeStyle({ backgroundColor: getStatus.value.color })
                    },
                    null,
                    6
                    /* CLASS, STYLE */
                  )) : createCommentVNode("v-if", true),
                  createTextVNode(
                    " " + toDisplayString(getStatus.value.label),
                    1
                    /* TEXT */
                  )
                ],
                16
                /* FULL_PROPS */
              )
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "tag" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 9 },
            [
              createCommentVNode(" \u6807\u7B7E "),
              createVNode(
                unref(ElTag),
                mergeProps({ class: "plus-display-item" }, customFieldProps.value),
                createSlots({
                  default: withCtx(() => [
                    createTextVNode(
                      " " + toDisplayString(displayValue.value),
                      1
                      /* TEXT */
                    )
                  ]),
                  _: 2
                  /* DYNAMIC */
                }, [
                  renderList(_ctx.column.fieldSlots, (fieldSlot, key) => {
                    return {
                      name: key,
                      fn: withCtx((data) => [
                        (openBlock(), createBlock(
                          resolveDynamicComponent(fieldSlot),
                          normalizeProps(guardReactiveProps(data)),
                          null,
                          16
                          /* FULL_PROPS */
                        ))
                      ])
                    };
                  })
                ]),
                1040
                /* FULL_PROPS, DYNAMIC_SLOTS */
              )
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "progress" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 10 },
            [
              createCommentVNode(" \u8FDB\u5EA6\u6761 "),
              createVNode(unref(ElProgress), mergeProps({
                class: "plus-display-item",
                percentage: displayValue.value
              }, customFieldProps.value), createSlots({
                _: 2
                /* DYNAMIC */
              }, [
                renderList(_ctx.column.fieldSlots, (fieldSlot, key) => {
                  return {
                    name: key,
                    fn: withCtx((data) => [
                      (openBlock(), createBlock(
                        resolveDynamicComponent(fieldSlot),
                        normalizeProps(guardReactiveProps(data)),
                        null,
                        16
                        /* FULL_PROPS */
                      ))
                    ])
                  };
                })
              ]), 1040, ["percentage"])
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "copy" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 11 },
            [
              createCommentVNode(" \u590D\u5236 "),
              createElementVNode("span", _hoisted_2, [
                createTextVNode(
                  toDisplayString(displayValue.value) + " ",
                  1
                  /* TEXT */
                ),
                createVNode(
                  unref(ElIcon),
                  mergeProps({
                    size: "16",
                    class: "plus-display-item__icon__copy"
                  }, customFieldProps.value, {
                    onClick: _cache[1] || (_cache[1] = ($event) => handelClickCopy(_ctx.column, subRow.value))
                  }),
                  {
                    default: withCtx(() => [
                      !subRow.value.isCopy ? (openBlock(), createBlock(unref(DocumentCopy), { key: 0 })) : (openBlock(), createBlock(unref(Select), { key: 1 }))
                    ]),
                    _: 1
                    /* STABLE */
                  },
                  16
                  /* FULL_PROPS */
                )
              ])
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : _ctx.column.valueType === "code" ? (openBlock(), createElementBlock(
            Fragment,
            { key: 12 },
            [
              createCommentVNode(" \u4EE3\u7801\u5757 "),
              createElementVNode(
                "pre",
                mergeProps({ class: "plus-display-item plus-display-item__pre" }, customFieldProps.value),
                "      " + toDisplayString(displayValue.value) + "\n  ",
                17
                /* TEXT, FULL_PROPS */
              )
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          )) : (openBlock(), createElementBlock(
            Fragment,
            { key: 13 },
            [
              createCommentVNode(" \u6CA1\u6709format "),
              createElementVNode(
                "span",
                mergeProps({ class: "plus-display-item" }, customFieldProps.value),
                toDisplayString(displayValue.value),
                17
                /* TEXT, FULL_PROPS */
              )
            ],
            2112
            /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
          ))
        ],
        2112
        /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
      );
    };
  }
});

export { _sfc_main as default };
