// @howdyjs/mouse-menu@2.1.3/dist/index.umd.js
(function(s,e){typeof exports=="object"&&typeof module<"u"?e(exports,require("vue")):typeof define=="function"&&define.amd?define(["exports","vue"],e):(s=typeof globalThis<"u"?globalThis:s||self,e(s.HowdyMouseMenu={},s.Vue))})(this,function(s,e){"use strict";const T=function(n){let m=Array.isArray(n)?[]:{};if(n&&typeof n=="object")for(let t in n)n.hasOwnProperty(t)&&(n[t]&&typeof n[t]=="object"?m[t]=T(n[t]):m[t]=n[t]);return m};var b=e.defineComponent({name:"MouseMenu",props:{appendToBody:{type:Boolean,default:!0},menuWidth:{type:Number,default:200},menuList:{type:Array,required:!0},menuHiddenFn:{type:Function},hasIcon:{type:Boolean,default:!1},iconType:{type:String,default:"font-icon"},menuWrapperCss:Object,menuItemCss:Object,el:{type:Object,required:!0},params:{type:[String,Number,Array,Object]},useLongPressInMobile:Boolean,longPressDuration:Number,longPressPreventDefault:[Function,Boolean],injectCloseListener:{type:Boolean,default:!0},customClass:String,disabled:{type:Function}},emits:["open","close"],setup(n,{emit:m}){const t=e.ref(0),o=e.ref(0),r=e.ref(!1),f=e.ref(0),_=e.ref(0),p=e.ref(!1),l=e.ref(null),v=e.ref([]),E=e.computed(()=>n.menuList.some(a=>a.children&&a.children.length>0)),k=e.ref(10),y=e.ref();e.watch(p,async a=>{if(a){await e.nextTick();let d=y.value;n.menuWrapperCss&&Object.keys(n.menuWrapperCss).map(u=>{d.style.setProperty(`--menu-${u}`,n.menuWrapperCss&&n.menuWrapperCss[u])}),n.menuItemCss&&Object.keys(n.menuItemCss).map(u=>{d.style.setProperty(`--menu-item-${u}`,n.menuItemCss&&n.menuItemCss[u])});let c=n.menuItemCss?.arrowSize?.match(/\d+/);c?k.value=~~c[0]||10:k.value=10,d.style.setProperty("--menu-item-arrowRealSize",k.value/2+"px"),m("open",n.params,l.value,n.el)}else m("close",n.params,l.value,n.el)});const Z=(a,d)=>{a.disabled||(a.fn&&typeof a.fn=="function"&&a.fn(n.params,l.value,n.el,d),p.value=!1)},ee=(a,d)=>{a.disabled||(a.fn&&typeof a.fn=="function"&&!a.disabled&&(a.fn(n.params,l.value,n.el,d),r.value=!1),p.value=!1)},ne=(a,d)=>{if(d.children&&!d.disabled){r.value=!0;const c=a.currentTarget;if(!c)return;const{offsetWidth:u}=c,i=c.querySelector(".__menu__sub__wrapper");if(!i)return;const{offsetWidth:w,offsetHeight:I}=i,{innerWidth:oe,innerHeight:D}=window,{top:P,left:L}=c.getBoundingClientRect();L+u+w>oe-5?t.value=L-w+5:t.value=L+u,P+I>D-5?o.value=D-I:o.value=P+5}},N=(a,d,c,u)=>a.map(i=>(i.children&&(i.children=N(i.children,d,c,u)),i.label&&typeof i.label=="function"&&(i.label=i.label(u,d,c)),i.tips&&typeof i.tips=="function"&&(i.tips=i.tips(u,d,c)),i.icon&&typeof i.icon=="function"&&(i.icon=i.icon(u,d,c)),i.hidden&&typeof i.hidden=="function"&&(i.hidden=i.hidden(u,d,c)),i.disabled&&typeof i.disabled=="function"&&(i.disabled=i.disabled(u,d,c)),i)),te=async(a=0,d=0)=>{if(l.value=document.elementFromPoint(a-1,d-1),n.menuHiddenFn?p.value=!n.menuHiddenFn(n.params,l.value,n.el):p.value=!0,!p.value)return;v.value=T(n.menuList),v.value=N(v.value,l.value,n.el,n.params),await e.nextTick();const{innerWidth:c,innerHeight:u}=window,i=y.value.offsetHeight,w=n.menuWidth||200;_.value=a+w+1>c?c-w-5:a+1,f.value=d+i+1>u?u-i-5:d+1},ie=()=>{p.value=!1},x=e.computed(()=>n.useLongPressInMobile&&"ontouchstart"in window?"touchstart":"mousedown"),S=a=>{y.value&&!y.value.contains(a.currentTarget)&&(p.value=!1,document.oncontextmenu=null)};return e.watch(()=>n.injectCloseListener,a=>{a?document.addEventListener(x.value,S):document.removeEventListener(x.value,S)},{immediate:!0}),e.onUnmounted(()=>{document.removeEventListener(x.value,S)}),{subLeft:t,subTop:o,hoverFlag:r,menuTop:f,menuLeft:_,showMenu:p,clickDomEl:l,calcMenuList:v,arrowSize:k,hasSubMenu:E,MenuWrapper:y,handleMenuItemClick:Z,handleSubMenuItemClick:ee,handleMenuMouseEnter:ne,show:te,close:ie,clickEventKey:x}}});const V=["onMouseenter"],j={key:0,class:"__menu__item-icon"},H=["innerHTML"],W={class:"__menu__item-label"},$={class:"__menu__item-tips"},A={class:"__menu__item-arrow-after"},R={class:"__menu__sub__item-label"},O={class:"__menu__sub__item-tips"};function q(n,m,t,o,r,f){return e.openBlock(),e.createBlock(e.Teleport,{to:"body",disabled:!n.appendToBody},[n.showMenu?(e.openBlock(),e.createElementBlock("div",{key:0,ref:"MenuWrapper",class:e.normalizeClass(["__menu__wrapper",n.customClass]),style:e.normalizeStyle({width:`${n.menuWidth}px`,top:`${n.menuTop}px`,left:`${n.menuLeft}px`})},[(e.openBlock(!0),e.createElementBlock(e.Fragment,null,e.renderList(n.calcMenuList,(_,p)=>(e.openBlock(),e.createElementBlock(e.Fragment,null,[!_.hidden&&!_.line?(e.openBlock(),e.createElementBlock("div",e.mergeProps({key:p,class:["__menu__item",_.disabled&&"disabled",_.customClass]},{[e.toHandlerKey(n.clickEventKey)]:e.withModifiers(l=>n.handleMenuItemClick(_,l),["stop"])},{onMouseenter:l=>n.handleMenuMouseEnter(l,_)}),[n.hasIcon?(e.openBlock(),e.createElementBlock("div",j,[n.iconType==="font-icon"?e.withDirectives((e.openBlock(),e.createElementBlock("i",{key:0,class:e.normalizeClass(_.icon)},null,2)),[[e.vShow,_.icon]]):n.iconType==="svg-icon"?e.withDirectives((e.openBlock(),e.createElementBlock("div",{key:1,class:"__menu__item-icon-svg",innerHTML:_.icon},null,8,H)),[[e.vShow,_.icon]]):n.iconType==="vnode-icon"?(e.openBlock(),e.createBlock(e.resolveDynamicComponent(_.icon),{key:2})):e.createCommentVNode("v-if",!0)])):e.createCommentVNode("v-if",!0),e.createElementVNode("span",W,e.toDisplayString(_.label),1),e.createElementVNode("span",$,e.toDisplayString(_.tips||""),1),n.hasSubMenu?(e.openBlock(),e.createElementBlock("span",{key:1,class:e.normalizeClass(["__menu__item-arrow",{show:n.hasSubMenu&&_.children}]),style:e.normalizeStyle({width:n.arrowSize+"px",height:n.arrowSize+"px"})},[e.withDirectives(e.createElementVNode("span",A,null,512),[[e.vShow,n.hasSubMenu&&_.children]])],6)):e.createCommentVNode("v-if",!0),_.children&&_.children.length>0?e.withDirectives((e.openBlock(),e.createElementBlock("div",{key:2,class:"__menu__sub__wrapper",style:e.normalizeStyle({top:`${n.subTop}px`,left:`${n.subLeft}px`})},[(e.openBlock(!0),e.createElementBlock(e.Fragment,null,e.renderList(_.children,(l,v)=>(e.openBlock(),e.createElementBlock(e.Fragment,null,[!l.hidden&&!l.line?(e.openBlock(),e.createElementBlock("div",e.mergeProps({key:v,class:["__menu__sub__item",l.disabled&&"disabled",l.customClass]},{[e.toHandlerKey(n.clickEventKey)]:e.withModifiers(E=>n.handleSubMenuItemClick(l,E),["stop"])}),[e.createElementVNode("span",R,e.toDisplayString(l.label),1),e.createElementVNode("span",O,e.toDisplayString(l.tips||""),1)],16)):e.createCommentVNode("v-if",!0),l.line?(e.openBlock(),e.createElementBlock("div",{key:v,class:"__menu__line"})):e.createCommentVNode("v-if",!0)],64))),256))],4)),[[e.vShow,n.hoverFlag]]):e.createCommentVNode("v-if",!0)],16,V)):e.createCommentVNode("v-if",!0),!_.hidden&&_.line?(e.openBlock(),e.createElementBlock("div",{key:p,class:"__menu__line"})):e.createCommentVNode("v-if",!0)],64))),256))],6)):e.createCommentVNode("v-if",!0)],8,["disabled"])}function K(n,m){m===void 0&&(m={});var t=m.insertAt;if(!(!n||typeof document>"u")){var o=document.head||document.getElementsByTagName("head")[0],r=document.createElement("style");r.type="text/css",t==="top"&&o.firstChild?o.insertBefore(r,o.firstChild):o.appendChild(r),r.styleSheet?r.styleSheet.cssText=n:r.appendChild(document.createTextNode(n))}}var U=`.__menu__mask[data-v-3d21bc0a] {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  z-index: 999;
}
.__menu__wrapper[data-v-3d21bc0a] {
  --menu-background: #c8f2f0;
  --menu-boxShadow: 0 1px 5px #888;
  --menu-padding: 5px 0;
  --menu-borderRadius: 0;
  --menu-item-height: 30px;
  --menu-item-padding: 0 10px;
  --menu-item-iconSize: 20px;
  --menu-item-iconFontSize: 14px;
  --menu-item-iconColor: #484852;
  --menu-item-labelColor: #484852;
  --menu-item-labelFontSize: 14px;
  --menu-item-tipsColor: #889;
  --menu-item-tipsFontSize: 12px;
  --menu-item-arrowColor: #484852;
  --menu-item-disabledColor: #bcc;
  --menu-item-hoverBackground: rgba(255, 255, 255, 0.8);
  --menu-item-hoverIconColor: inherit;
  --menu-item-hoverLabelColor: inherit;
  --menu-item-hoverTipsColor: inherit;
  --menu-item-hoverArrowColor: inherit;
  --menu-lineColor: #ccc;
  --menu-lineMargin: 5px 0;
}
.__menu__wrapper[data-v-3d21bc0a] {
  position: fixed;
  width: 200px;
  background: var(--menu-background);
  box-shadow: var(--menu-boxShadow);
  padding: var(--menu-padding);
  border-radius: var(--menu-borderRadius);
  z-index: 99999;
}
.__menu__line[data-v-3d21bc0a],
.__menu__sub__line[data-v-3d21bc0a] {
  border-top: 1px solid var(--menu-lineColor);
  margin: var(--menu-lineMargin);
}
.__menu__item[data-v-3d21bc0a],
.__menu__sub__item[data-v-3d21bc0a] {
  display: flex;
  height: var(--menu-item-height);
  align-items: center;
  cursor: pointer;
  padding: var(--menu-item-padding);
}
.__menu__item .__menu__item-icon[data-v-3d21bc0a],
.__menu__sub__item .__menu__item-icon[data-v-3d21bc0a] {
  display: flex;
  align-items: center;
  justify-content: center;
  color: var(--menu-item-iconColor);
  width: var(--menu-item-iconSize);
  height: var(--menu-item-iconSize);
}
.__menu__item .__menu__item-icon i[data-v-3d21bc0a],
.__menu__sub__item .__menu__item-icon i[data-v-3d21bc0a] {
  font-size: var(--menu-item-iconFontSize);
}
.__menu__item .__menu__item-icon .__menu__item-icon-svg[data-v-3d21bc0a],
.__menu__sub__item .__menu__item-icon .__menu__item-icon-svg[data-v-3d21bc0a] {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
}
.__menu__item .__menu__item-label[data-v-3d21bc0a],
.__menu__item .__menu__sub__item-label[data-v-3d21bc0a],
.__menu__sub__item .__menu__item-label[data-v-3d21bc0a],
.__menu__sub__item .__menu__sub__item-label[data-v-3d21bc0a] {
  width: 100%;
  max-height: 100%;
  text-overflow: ellipsis;
  white-space: nowrap;
  flex: 1;
  font-size: var(--menu-item-labelFontSize);
  color: var(--menu-item-labelColor);
  margin-right: 5px;
  overflow: hidden;
}
.__menu__item .__menu__item-tips[data-v-3d21bc0a],
.__menu__item .__menu__sub__item-tips[data-v-3d21bc0a],
.__menu__sub__item .__menu__item-tips[data-v-3d21bc0a],
.__menu__sub__item .__menu__sub__item-tips[data-v-3d21bc0a] {
  font-size: var(--menu-item-tipsFontSize);
  color: var(--menu-item-tipsColor);
}
.__menu__item .__menu__item-arrow[data-v-3d21bc0a],
.__menu__sub__item .__menu__item-arrow[data-v-3d21bc0a] {
  width: 10px;
  height: 10px;
  margin-left: 5px;
  position: relative;
}
.__menu__item.disabled[data-v-3d21bc0a],
.__menu__sub__item.disabled[data-v-3d21bc0a] {
  cursor: not-allowed;
}
.__menu__item.disabled .__menu__item-icon[data-v-3d21bc0a],
.__menu__item.disabled .__menu__item-label[data-v-3d21bc0a],
.__menu__item.disabled .__menu__sub__item-label[data-v-3d21bc0a],
.__menu__item.disabled .__menu__item-tips[data-v-3d21bc0a],
.__menu__item.disabled .__menu__sub__item-tips[data-v-3d21bc0a],
.__menu__sub__item.disabled .__menu__item-icon[data-v-3d21bc0a],
.__menu__sub__item.disabled .__menu__item-label[data-v-3d21bc0a],
.__menu__sub__item.disabled .__menu__sub__item-label[data-v-3d21bc0a],
.__menu__sub__item.disabled .__menu__item-tips[data-v-3d21bc0a],
.__menu__sub__item.disabled .__menu__sub__item-tips[data-v-3d21bc0a] {
  color: var(--menu-item-disabledColor);
}
.__menu__item.disabled .__menu__item-arrow .__menu__item-arrow-after[data-v-3d21bc0a],
.__menu__sub__item.disabled .__menu__item-arrow .__menu__item-arrow-after[data-v-3d21bc0a] {
  border-left: var(--menu-item-arrowRealSize) solid var(--menu-item-disabledColor);
}
.__menu__item[data-v-3d21bc0a]:not(.disabled):hover {
  background: var(--menu-item-hoverBackground);
}
.__menu__item:not(.disabled):hover .__menu__item-icon[data-v-3d21bc0a] {
  color: var(--menu-item-hoverIconColor);
}
.__menu__item:not(.disabled):hover .__menu__item-label[data-v-3d21bc0a] {
  color: var(--menu-item-hoverLabelColor);
}
.__menu__item:not(.disabled):hover .__menu__item-tips[data-v-3d21bc0a] {
  color: var(--menu-item-hoverTipsColor);
}
.__menu__item:not(.disabled):hover .__menu__item-arrow[data-v-3d21bc0a] {
  color: var(--menu-item-hoverArrowColor);
}
.__menu__sub__item[data-v-3d21bc0a]:not(.disabled):hover {
  background: var(--menu-item-hoverBackground);
}
.__menu__sub__item:not(.disabled):hover .__menu__sub__item-label[data-v-3d21bc0a] {
  color: var(--menu-item-hoverLabelColor);
}
.__menu__sub__item:not(.disabled):hover .__menu__sub__item-tips[data-v-3d21bc0a] {
  color: var(--menu-item-hoverTipsColor);
}
.__menu__item-icon[data-v-3d21bc0a] {
  width: 20px;
  height: 20px;
  text-align: center;
  line-height: 20px;
  margin-right: 4px;
}
.__menu__item-arrow.show .__menu__item-arrow-after[data-v-3d21bc0a] {
  position: absolute;
  width: 0;
  height: 0;
  left: 8px;
  border-left: var(--menu-item-arrowRealSize) solid var(--menu-item-arrowColor);
  border-top: var(--menu-item-arrowRealSize) solid transparent;
  border-bottom: var(--menu-item-arrowRealSize) solid transparent;
}
.__menu__sub__wrapper[data-v-3d21bc0a] {
  position: fixed;
  visibility: hidden;
  width: 200px;
  background: var(--menu-background);
  box-shadow: var(--menu-boxShadow);
  padding: var(--menu-padding);
  border-radius: var(--menu-borderRadius);
}
.__menu__item:hover .__menu__sub__wrapper[data-v-3d21bc0a] {
  visibility: visible;
}`;K(U),b.render=q,b.__scopeId="data-v-3d21bc0a",b.__file="packages/mouse-menu/mouse-menu.vue";function X(n,m,t){let o=document.createElement(n);return o.setAttribute("class",m),t&&(o.innerText=t),o}b.install=n=>{n.component(b.name,b)};function M(n){const m="__mouse__menu__container";let t;document.querySelector(`.${m}`)?t=document.querySelector(`.${m}`):t=X("div",m);const o=e.createVNode(b,n);return e.render(o,t),document.body.appendChild(t),o.component?.proxy}s.MouseMenuCtx=void 0;let g,B,h;function Y(n,m,t=500,o){B=r=>{s.MouseMenuCtx&&s.MouseMenuCtx.close(),typeof o=="function"?o(r,n)&&r.preventDefault():typeof o=="boolean"&&o&&r.preventDefault(),g&&clearTimeout(g),g=window.setTimeout(()=>{m(r)},t)},h=()=>{clearTimeout(g)},n.addEventListener("touchstart",B),n.addEventListener("touchmove",h),n.addEventListener("touchend",h),n.addEventListener("touchcancel",h)}function z(n){n.removeEventListener("touchstart",B),n.removeEventListener("touchmove",h),n.removeEventListener("touchend",h),n.removeEventListener("touchcancel",h)}let C,F;const G=(n,m)=>{const{value:t}=m;if(t.menuList.length>0)C=o=>{if(typeof t.disabled=="function"&&t.disabled(t.params))return;o.preventDefault(),s.MouseMenuCtx=M({el:n,...t});const{x:r,y:f}=o;s.MouseMenuCtx.show(r,f)},n.removeEventListener("contextmenu",C),n.addEventListener("contextmenu",C),t.useLongPressInMobile&&"ontouchstart"in window&&(F=o=>{if(typeof t.disabled=="function"&&t.disabled(t.params))return;s.MouseMenuCtx=M({el:n,...t});const{touches:r}=o,{clientX:f,clientY:_}=r[0];s.MouseMenuCtx.show(f,_),document.onmousedown=null,n.onmousedown=null,setTimeout(()=>{document.onmousedown=()=>s.MouseMenuCtx.close(),n.onmousedown=()=>s.MouseMenuCtx.close()},500)},z(n),Y(n,F,t.longPressDuration||500,t.longPressPreventDefault));else throw new Error("At least set one menu list!")},J=n=>{n.removeEventListener("contextmenu",C),"touchstart"in window&&z(n)},Q={mounted:G,unmounted:J};s.CustomMouseMenu=M,s.MouseMenuDirective=Q,s.default=b,Object.defineProperty(s,"__esModule",{value:!0})});
