// element-plus@2.1.4/lib/tokens/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var breadcrumb = require('./breadcrumb.js');
var button = require('./button.js');
var collapse = require('./collapse.js');
var configProvider = require('./config-provider.js');
var dialog = require('./dialog.js');
var form = require('./form.js');
var pagination = require('./pagination.js');
var radio = require('./radio.js');
var row = require('./row.js');
var scrollbar = require('./scrollbar.js');
var tabs = require('./tabs.js');
var upload = require('./upload.js');



exports.breadcrumbKey = breadcrumb.breadcrumbKey;
exports.buttonGroupContextKey = button.buttonGroupContextKey;
exports.collapseContextKey = collapse.collapseContextKey;
exports.configProviderContextKey = configProvider.configProviderContextKey;
exports.dialogInjectionKey = dialog.dialogInjectionKey;
exports.formContextKey = form.formContextKey;
exports.formItemContextKey = form.formItemContextKey;
exports.elPaginationKey = pagination.elPaginationKey;
exports.radioGroupKey = radio.radioGroupKey;
exports.rowContextKey = row.rowContextKey;
exports.scrollbarContextKey = scrollbar.scrollbarContextKey;
exports.tabsRootContextKey = tabs.tabsRootContextKey;
exports.uploadContextKey = upload.uploadContextKey;
