// plus-pro-components/es/components/page/src/index.vue2.mjs
import { defineComponent, computed, ref, reactive, useSlots, h, openBlock, createElementBlock, createBlock, resolveDynamicComponent, withCtx, createVNode, unref, mergeProps, createSlots, renderSlot, normalizeProps, guardReactiveProps, createCommentVNode, renderList } from 'vue';
import { PlusSearch } from '../../search/index.mjs';
import { PlusTable } from '../../table/index.mjs';
import { ElCard } from 'element-plus';
import '../../../hooks/index.mjs';
import { filterSlots, getTableCellSlotName, getTableHeaderSlotName, getFieldSlotName } from '../../utils/index.mjs';
import '../../../constants/index.mjs';
import { DefaultPageInfo, DefaultPageSizeList } from '../../../constants/page.mjs';
import { useTable } from '../../../hooks/useTable.mjs';

const _hoisted_1 = { class: "plus-page" };
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "PlusPage"
  },
  __name: "index",
  props: {
    columns: { type: Array, required: true, default: () => [] },
    request: { type: Function, required: true },
    search: { type: [Boolean, Object], required: false, default: () => ({}) },
    table: { type: Object, required: false, default: () => ({}) },
    params: { type: null, required: false, default: () => ({}) },
    postData: { type: Function, required: false, default: void 0 },
    beforeSearchSubmit: { type: Function, required: false, default: void 0 },
    isCard: { type: Boolean, required: false, default: true },
    searchCardProps: { type: Object, required: false, default: () => ({}) },
    tableCardProps: { type: Object, required: false, default: () => ({}) },
    defaultPageInfo: { type: null, required: false, default: () => ({ ...DefaultPageInfo }) },
    defaultPageSizeList: { type: Array, required: false, default: () => DefaultPageSizeList },
    pagination: { type: null, required: false, default: () => ({}) }
  },
  emits: ["requestError", "search", "reset", "paginationChange"],
  setup(__props, { expose: __expose, emit }) {
    var _a;
    const props = __props;
    const PlusSearch$1 = PlusSearch;
    const PlusTable$1 = PlusTable;
    const computedDefaultPageInfo = computed(() => props.defaultPageInfo);
    const computedDefaultPageSizeList = computed(() => props.defaultPageSizeList);
    const { tableData, pageInfo, total, loadingStatus } = useTable(computedDefaultPageInfo);
    const plusSearchInstance = ref();
    const plusTableInstance = ref();
    const state = reactive({
      params: { ...(_a = props.search) == null ? void 0 : _a.defaultValues },
      values: {}
    });
    const slots = useSlots();
    const cellSlots = filterSlots(slots, getTableCellSlotName());
    const headerSlots = filterSlots(slots, getTableHeaderSlotName());
    const fieldSlots = filterSlots(slots, getFieldSlotName());
    const renderWrapper = () => {
      if (props.isCard) {
        return {
          search: h(ElCard, props.searchCardProps),
          table: h(ElCard, props.tableCardProps)
        };
      }
      return { search: h("div"), table: h("div") };
    };
    const getList = async () => {
      if (!props.request)
        return;
      try {
        loadingStatus.value = true;
        const { data, total: dataTotal } = await props.request({
          ...state.params,
          ...pageInfo.value,
          ...props.params
        });
        const list = props.postData && props.postData(data) || data;
        tableData.value = list || [];
        total.value = dataTotal || list.length;
      } catch (error) {
        emit("requestError", error);
      }
      loadingStatus.value = false;
    };
    getList();
    const handlePaginationChange = (_pageInfo) => {
      pageInfo.value = _pageInfo;
      getList();
      emit("paginationChange", _pageInfo);
    };
    const handleSearch = (values) => {
      const data = props.beforeSearchSubmit && props.beforeSearchSubmit(values) || values;
      state.params = data;
      getList();
      emit("search", state.params);
    };
    const handleRest = (values) => {
      state.params = { ...values };
      pageInfo.value.page = 1;
      getList();
      emit("reset", state.params);
    };
    const handleRefresh = () => {
      getList();
    };
    __expose({
      plusSearchInstance,
      plusTableInstance,
      getList,
      handleRest
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        _ctx.search ? (openBlock(), createBlock(resolveDynamicComponent(renderWrapper().search), { key: 0 }, {
          default: withCtx(() => [
            createVNode(unref(PlusSearch$1), mergeProps({
              ref_key: "plusSearchInstance",
              ref: plusSearchInstance
            }, _ctx.search, {
              modelValue: state.params,
              "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => state.params = $event),
              columns: _ctx.columns,
              "search-loading": unref(loadingStatus),
              onSearch: handleSearch,
              onReset: handleRest
            }), createSlots({
              _: 2
              /* DYNAMIC */
            }, [
              _ctx.$slots["search-footer"] ? {
                name: "footer",
                fn: withCtx((data) => [
                  renderSlot(_ctx.$slots, "search-footer", normalizeProps(guardReactiveProps(data)))
                ]),
                key: "0"
              } : void 0
            ]), 1040, ["modelValue", "columns", "search-loading"])
          ]),
          _: 3
          /* FORWARDED */
        })) : createCommentVNode("v-if", true),
        (openBlock(), createBlock(resolveDynamicComponent(renderWrapper().table), { class: "plus-page__table_wrapper" }, {
          default: withCtx(() => [
            createVNode(unref(PlusTable$1), mergeProps({
              ref_key: "plusTableInstance",
              ref: plusTableInstance,
              "title-bar": { refresh: true }
            }, _ctx.table, {
              "table-data": unref(tableData),
              "loading-status": unref(loadingStatus),
              columns: _ctx.columns,
              pagination: {
                ..._ctx.pagination,
                total: unref(total),
                modelValue: unref(pageInfo),
                pageSizeList: computedDefaultPageSizeList.value
              },
              onPaginationChange: handlePaginationChange,
              onRefresh: handleRefresh
            }), createSlots({
              _: 2
              /* DYNAMIC */
            }, [
              renderList(unref(headerSlots), (_, key) => {
                return {
                  name: key,
                  fn: withCtx((data) => [
                    renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                  ])
                };
              }),
              renderList(unref(cellSlots), (_, key) => {
                return {
                  name: key,
                  fn: withCtx((data) => [
                    renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                  ])
                };
              }),
              renderList(unref(fieldSlots), (_, key) => {
                return {
                  name: key,
                  fn: withCtx((data) => [
                    renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                  ])
                };
              }),
              _ctx.$slots["table-title"] ? {
                name: "title",
                fn: withCtx(() => [
                  renderSlot(_ctx.$slots, "table-title")
                ]),
                key: "0"
              } : void 0,
              _ctx.$slots["table-toolbar"] ? {
                name: "toolbar",
                fn: withCtx(() => [
                  renderSlot(_ctx.$slots, "table-toolbar")
                ]),
                key: "1"
              } : void 0,
              _ctx.$slots["table-expand"] ? {
                name: "expand",
                fn: withCtx(() => [
                  renderSlot(_ctx.$slots, "table-expand")
                ]),
                key: "2"
              } : void 0,
              _ctx.$slots["table-append"] ? {
                name: "append",
                fn: withCtx(() => [
                  renderSlot(_ctx.$slots, "table-append")
                ]),
                key: "3"
              } : void 0,
              _ctx.$slots["table-empty"] ? {
                name: "empty",
                fn: withCtx(() => [
                  renderSlot(_ctx.$slots, "table-empty")
                ]),
                key: "4"
              } : void 0
            ]), 1040, ["table-data", "loading-status", "columns", "pagination"])
          ]),
          _: 3
          /* FORWARDED */
        }))
      ]);
    };
  }
});

export { _sfc_main as default };
