// plus-pro-components/lib/components/table/src/index.vue2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var index$1 = require('../../pagination/index.js');
require('../../../constants/index.js');
var elementPlus = require('element-plus');
var index = require('../../utils/index.js');
var tableActionBar = require('./table-action-bar.vue.js');
var tableColumn = require('./table-column.vue.js');
var tableColumnIndex = require('./table-column-index.vue.js');
var tableColumnDragSort = require('./table-column-drag-sort.vue.js');
var tableTitleBar = require('./table-title-bar.vue.js');
var page = require('../../../constants/page.js');
var form = require('../../../constants/form.js');

const _hoisted_1 = { class: "plus-table" };
const _hoisted_2 = { class: "plus-table-expand-col" };
var _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    name: "PlusTable",
    inheritAttrs: false
  },
  __name: "index",
  props: {
    defaultSize: { type: null, required: false, default: "default" },
    pagination: { type: [Boolean, Object], required: false, default: false },
    actionBar: { type: [Boolean, Object], required: false, default: false },
    hasIndexColumn: { type: Boolean, required: false, default: false },
    titleBar: { type: [Boolean, Object], required: false, default: true },
    isSelection: { type: Boolean, required: false, default: false },
    hasExpand: { type: Boolean, required: false, default: false },
    loadingStatus: { type: Boolean, required: false, default: false },
    height: { type: String, required: false },
    tableData: { type: Array, required: true, default: () => [] },
    columns: { type: Array, required: true, default: () => [] },
    headerCellStyle: { type: Object, required: false, default: () => ({}) },
    rowKey: { type: String, required: false, default: "id" },
    dragSortable: { type: [Boolean, Object], required: false, default: false },
    dragSortableTableColumnProps: { type: null, required: false, default: () => ({}) },
    indexTableColumnProps: { type: null, required: false, default: () => ({}) },
    selectionTableColumnProps: { type: null, required: false, default: () => ({
      width: 40
    }) },
    expandTableColumnProps: { type: null, required: false, default: () => ({}) },
    indexContentStyle: { type: [Object, Function], required: false, default: () => ({}) }
  },
  emits: ["paginationChange", "clickAction", "clickActionConfirmCancel", "dragSortEnd", "formChange", "refresh"],
  setup(__props, { expose: __expose, emit }) {
    var _a;
    const props = __props;
    const PlusTableActionBar = tableActionBar.default;
    const PlusTableColumn = tableColumn.default;
    const subColumns = vue.ref([]);
    const tableInstance = vue.shallowRef(null);
    const state = vue.reactive({
      subPageInfo: {
        ...((_a = props.pagination) == null ? void 0 : _a.modelValue) || page.DefaultPageInfo
      },
      size: props.defaultSize
    });
    const slots = vue.useSlots();
    const cellSlots = index.filterSlots(slots, index.getTableCellSlotName());
    const headerSlots = index.filterSlots(slots, index.getTableHeaderSlotName());
    const fieldSlots = index.filterSlots(slots, index.getFieldSlotName());
    const extraSlots = index.filterSlots(slots, index.getExtraSlotName());
    const formRefs = vue.shallowRef({});
    vue.provide(form.TableFormRefInjectionKey, formRefs);
    vue.watch(
      () => props.columns,
      (val) => {
        subColumns.value = val.filter((item) => vue.unref(item.hideInTable) !== true);
      },
      {
        deep: true,
        immediate: true
      }
    );
    const handlePaginationChange = () => {
      emit("paginationChange", { ...state.subPageInfo });
    };
    const handleAction = (res) => {
      const { row, buttonRow, index, e } = res;
      emit("clickAction", { row, buttonRow, index, e, formRefs: formRefs.value[index] });
    };
    const handleClickActionConfirmCancel = (res) => {
      const { row, buttonRow, index, e } = res;
      emit("clickActionConfirmCancel", {
        row,
        buttonRow,
        index,
        e,
        formRefs: formRefs.value[index]
      });
    };
    const handleFilterTableConfirm = (data) => {
      subColumns.value = data.filter((item) => vue.unref(item.hideInTable) !== true);
    };
    const handleClickDensity = (data) => {
      state.size = data;
    };
    const handleDragSortEnd = (newIndex, oldIndex) => {
      emit("dragSortEnd", newIndex, oldIndex);
    };
    const handleRefresh = () => {
      emit("refresh");
    };
    const handleFormChange = (data) => {
      emit("formChange", data);
    };
    const { subPageInfo, size } = vue.toRefs(state);
    __expose({
      formRefs,
      tableInstance
    });
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createElementBlock("div", _hoisted_1, [
        _ctx.titleBar ? (vue.openBlock(), vue.createBlock(tableTitleBar.default, {
          key: 0,
          columns: _ctx.columns,
          "default-size": vue.unref(size),
          "title-bar": _ctx.titleBar,
          "change-columns": subColumns.value,
          onClickDensity: handleClickDensity,
          onFilterTable: handleFilterTableConfirm,
          onRefresh: handleRefresh
        }, {
          title: vue.withCtx(() => [
            vue.renderSlot(_ctx.$slots, "title")
          ]),
          toolbar: vue.withCtx(() => [
            vue.renderSlot(_ctx.$slots, "toolbar")
          ]),
          _: 3
          /* FORWARDED */
        }, 8, ["columns", "default-size", "title-bar", "change-columns"])) : vue.createCommentVNode("v-if", true),
        vue.withDirectives((vue.openBlock(), vue.createBlock(vue.unref(elementPlus.ElTable), vue.mergeProps({
          ref_key: "tableInstance",
          ref: tableInstance,
          "reserve-selection": true,
          data: _ctx.tableData,
          border: true,
          height: _ctx.height,
          "header-cell-style": _ctx.headerCellStyle,
          size: vue.unref(size),
          "row-key": _ctx.rowKey,
          "highlight-current-row": "",
          "scrollbar-always-on": ""
        }, _ctx.$attrs), {
          append: vue.withCtx(() => [
            vue.renderSlot(_ctx.$slots, "append")
          ]),
          empty: vue.withCtx(() => [
            vue.renderSlot(_ctx.$slots, "empty")
          ]),
          default: vue.withCtx(() => {
            var _a2;
            return [
              _ctx.isSelection ? (vue.openBlock(), vue.createBlock(
                vue.unref(elementPlus.ElTableColumn),
                vue.mergeProps({
                  key: "selection",
                  type: "selection"
                }, _ctx.selectionTableColumnProps),
                null,
                16
                /* FULL_PROPS */
              )) : vue.createCommentVNode("v-if", true),
              _ctx.hasIndexColumn ? (vue.openBlock(), vue.createBlock(tableColumnIndex.default, {
                key: 1,
                "index-content-style": _ctx.indexContentStyle,
                "index-table-column-props": _ctx.indexTableColumnProps,
                "page-info": (_a2 = _ctx.pagination) == null ? void 0 : _a2.modelValue
              }, null, 8, ["index-content-style", "index-table-column-props", "page-info"])) : vue.createCommentVNode("v-if", true),
              _ctx.dragSortable ? (vue.openBlock(), vue.createBlock(tableColumnDragSort.default, {
                key: 2,
                sortable: _ctx.dragSortable,
                "drag-sortable-table-column-props": _ctx.dragSortableTableColumnProps,
                "table-instance": tableInstance.value,
                onDragSortEnd: handleDragSortEnd
              }, null, 8, ["sortable", "drag-sortable-table-column-props", "table-instance"])) : vue.createCommentVNode("v-if", true),
              _ctx.hasExpand ? (vue.openBlock(), vue.createBlock(
                vue.unref(elementPlus.ElTableColumn),
                vue.mergeProps({
                  key: 3,
                  type: "expand"
                }, _ctx.expandTableColumnProps),
                {
                  default: vue.withCtx(({ row, $index }) => [
                    vue.createElementVNode("div", _hoisted_2, [
                      vue.renderSlot(_ctx.$slots, "expand", {
                        row,
                        index: $index
                      })
                    ])
                  ]),
                  _: 3
                  /* FORWARDED */
                },
                16
                /* FULL_PROPS */
              )) : vue.createCommentVNode("v-if", true),
              vue.createVNode(vue.unref(PlusTableColumn), {
                columns: subColumns.value,
                onFormChange: handleFormChange
              }, vue.createSlots({
                _: 2
                /* DYNAMIC */
              }, [
                vue.renderList(vue.unref(headerSlots), (_, key) => {
                  return {
                    name: key,
                    fn: vue.withCtx((data) => [
                      vue.renderSlot(_ctx.$slots, key, vue.normalizeProps(vue.guardReactiveProps(data)))
                    ])
                  };
                }),
                vue.renderList(vue.unref(cellSlots), (_, key) => {
                  return {
                    name: key,
                    fn: vue.withCtx((data) => [
                      vue.renderSlot(_ctx.$slots, key, vue.normalizeProps(vue.guardReactiveProps(data)))
                    ])
                  };
                }),
                vue.renderList(vue.unref(fieldSlots), (_, key) => {
                  return {
                    name: key,
                    fn: vue.withCtx((data) => [
                      vue.renderSlot(_ctx.$slots, key, vue.normalizeProps(vue.guardReactiveProps(data)))
                    ])
                  };
                }),
                vue.renderList(vue.unref(extraSlots), (_, key) => {
                  return {
                    name: key,
                    fn: vue.withCtx((data) => [
                      vue.renderSlot(_ctx.$slots, key, vue.normalizeProps(vue.guardReactiveProps(data)))
                    ])
                  };
                })
              ]), 1032, ["columns"]),
              _ctx.actionBar ? (vue.openBlock(), vue.createBlock(
                vue.unref(PlusTableActionBar),
                vue.mergeProps({ key: 4 }, _ctx.actionBar, {
                  onClickAction: handleAction,
                  onClickActionConfirmCancel: handleClickActionConfirmCancel
                }),
                null,
                16
                /* FULL_PROPS */
              )) : vue.createCommentVNode("v-if", true)
            ];
          }),
          _: 3
          /* FORWARDED */
        }, 16, ["data", "height", "header-cell-style", "size", "row-key"])), [
          [vue.unref(elementPlus.vLoading), _ctx.loadingStatus]
        ]),
        vue.createCommentVNode(" \u5206\u9875 "),
        _ctx.pagination ? (vue.openBlock(), vue.createBlock(vue.unref(index$1.PlusPagination), vue.mergeProps({
          key: 1,
          modelValue: vue.unref(subPageInfo),
          "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => vue.isRef(subPageInfo) ? subPageInfo.value = $event : null)
        }, _ctx.pagination, { onChange: handlePaginationChange }), vue.createSlots({
          _: 2
          /* DYNAMIC */
        }, [
          _ctx.$slots["pagination-left"] ? {
            name: "pagination-left",
            fn: vue.withCtx(() => [
              vue.renderSlot(_ctx.$slots, "pagination-left")
            ]),
            key: "0"
          } : void 0,
          _ctx.$slots["pagination-right"] ? {
            name: "pagination-right",
            fn: vue.withCtx(() => [
              vue.renderSlot(_ctx.$slots, "pagination-right")
            ]),
            key: "1"
          } : void 0
        ]), 1040, ["modelValue"])) : vue.createCommentVNode("v-if", true)
      ]);
    };
  }
});

exports.default = _sfc_main;
