// plus-pro-components/lib/components/table/src/table-action-bar.vue2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var iconsVue = require('@element-plus/icons-vue');
var elementPlus = require('element-plus');
require('../../utils/index.js');
require('../../../hooks/index.js');
var useLocale = require('../../../hooks/useLocale.js');
var is = require('../../utils/is.js');

const _hoisted_1 = { class: "plus-table-action-bar__dropdown__link" };
var _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    name: "PlusTableActionBar"
  },
  __name: "table-action-bar",
  props: {
    label: { type: String, required: false, default: "" },
    fixed: { type: String, required: false, default: "right" },
    showNumber: { type: Number, required: false, default: 3 },
    type: { type: String, required: false, default: "link" },
    buttons: { type: Array, required: false, default: () => [] },
    width: { type: [String, Number], required: false, default: 200 },
    actionBarTableColumnProps: { type: Object, required: false, default: () => ({}) }
  },
  emits: ["clickAction", "clickActionConfirmCancel"],
  setup(__props, { emit }) {
    const props = __props;
    const { t } = useLocale.useLocale();
    const getSubButtons = (row, index) => {
      const data = props.buttons.filter((item) => {
        if (is.isFunction(item.show)) {
          const tempFunction = item.show;
          const isShow = tempFunction(row, index, item);
          return vue.unref(isShow) !== false;
        }
        return vue.unref(item.show) !== false;
      });
      const preButtons = data.slice(0, props.showNumber);
      const nextButtons = data.slice(props.showNumber);
      const showMore = data.length > props.showNumber;
      return {
        showMore,
        preButtons,
        nextButtons
      };
    };
    const render = (row, buttonRow, index) => {
      if (props.type === "icon") {
        return vue.h(
          elementPlus.ElTooltip,
          { placement: "top", content: vue.unref(buttonRow.text), ...buttonRow.tooltipProps },
          () => vue.h(
            elementPlus.ElIcon,
            {
              size: 16,
              ...buttonRow.props,
              onClick: (event) => handleClickAction(row, buttonRow, index, event)
            },
            () => buttonRow.icon ? vue.h(buttonRow.icon) : ""
          )
        );
      } else {
        const Tag = props.type === "button" ? elementPlus.ElButton : elementPlus.ElLink;
        const defaultProps = props.type === "link" ? { href: "javaScript:;" } : {};
        return vue.h(
          Tag,
          {
            size: "small",
            ...defaultProps,
            ...buttonRow.props,
            onClick: (event) => handleClickAction(row, buttonRow, index, event)
          },
          () => vue.unref(buttonRow.text)
        );
      }
    };
    const handleClickAction = (row, buttonRow, index, e) => {
      var _a, _b;
      const data = { row, buttonRow, index, e };
      if (buttonRow.confirm) {
        const message = t("plus.table.confirmToPerformThisOperation");
        let title = t("plus.table.prompt");
        let options = void 0;
        let appContext = null;
        if (is.isPlainObject(buttonRow.confirm) && typeof buttonRow.confirm !== "boolean") {
          const tempTitle = is.isFunction(buttonRow.confirm.title) ? buttonRow.confirm.title(data) : buttonRow.confirm.title;
          if (tempTitle) {
            title = tempTitle;
          }
          const tempMessage = is.isFunction(buttonRow.confirm.message) ? buttonRow.confirm.message(data) : buttonRow.confirm.message;
          if (tempMessage) {
            title = tempMessage;
          }
          options = (_a = buttonRow.confirm) == null ? void 0 : _a.options;
          appContext = (_b = buttonRow.confirm) == null ? void 0 : _b.appContext;
        }
        elementPlus.ElMessageBox.confirm(message, title, options, appContext).then(() => {
          emit("clickAction", data);
        }).catch(() => {
          emit("clickActionConfirmCancel", data);
        });
      } else {
        emit("clickAction", data);
      }
    };
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createBlock(vue.unref(elementPlus.ElTableColumn), vue.mergeProps({
        key: "actionBar",
        "class-name": "plus-table-action-bar",
        align: "center",
        label: _ctx.label || vue.unref(t)("plus.table.action"),
        fixed: _ctx.fixed || "right",
        width: _ctx.width || 200
      }, props.actionBarTableColumnProps), {
        default: vue.withCtx(({ row, $index }) => [
          vue.createCommentVNode(" \u663E\u793A\u51FA\u6765\u7684\u6309\u94AE "),
          (vue.openBlock(true), vue.createElementBlock(
            vue.Fragment,
            null,
            vue.renderList(getSubButtons(row, $index).preButtons, (buttonRow) => {
              return vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent(() => render(row, buttonRow, $index)), {
                key: buttonRow.text
              });
            }),
            128
            /* KEYED_FRAGMENT */
          )),
          vue.createCommentVNode(" \u9690\u85CF\u7684\u6309\u94AE "),
          getSubButtons(row, $index).showMore ? (vue.openBlock(), vue.createBlock(
            vue.unref(elementPlus.ElDropdown),
            {
              key: 0,
              trigger: "click",
              class: "plus-table-action-bar__dropdown"
            },
            {
              dropdown: vue.withCtx(() => [
                vue.createVNode(
                  vue.unref(elementPlus.ElDropdownMenu),
                  null,
                  {
                    default: vue.withCtx(() => [
                      (vue.openBlock(true), vue.createElementBlock(
                        vue.Fragment,
                        null,
                        vue.renderList(getSubButtons(row, $index).nextButtons, (buttonRow) => {
                          return vue.openBlock(), vue.createBlock(
                            vue.unref(elementPlus.ElDropdownItem),
                            {
                              key: vue.unref(buttonRow.text)
                            },
                            {
                              default: vue.withCtx(() => [
                                (vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent(() => render(row, buttonRow, $index))))
                              ]),
                              _: 2
                              /* DYNAMIC */
                            },
                            1024
                            /* DYNAMIC_SLOTS */
                          );
                        }),
                        128
                        /* KEYED_FRAGMENT */
                      ))
                    ]),
                    _: 2
                    /* DYNAMIC */
                  },
                  1024
                  /* DYNAMIC_SLOTS */
                )
              ]),
              default: vue.withCtx(() => [
                vue.createElementVNode("span", _hoisted_1, [
                  vue.createTextVNode(
                    vue.toDisplayString(vue.unref(t)("plus.table.more")) + " ",
                    1
                    /* TEXT */
                  ),
                  vue.createVNode(vue.unref(elementPlus.ElIcon), null, {
                    default: vue.withCtx(() => [
                      vue.createVNode(vue.unref(iconsVue.ArrowDownBold))
                    ]),
                    _: 1
                    /* STABLE */
                  })
                ])
              ]),
              _: 2
              /* DYNAMIC */
            },
            1024
            /* DYNAMIC_SLOTS */
          )) : vue.createCommentVNode("v-if", true)
        ]),
        _: 1
        /* STABLE */
      }, 16, ["label", "fixed", "width"]);
    };
  }
});

exports.default = _sfc_main;
