// plus-pro-components/es/components/form/src/index.vue2.mjs
import { defineComponent, ref, reactive, unref, computed, useSlots, watch, openBlock, createBlock, mergeProps, withCtx, renderSlot, createCommentVNode, createElementBlock, Fragment, renderList, createElementVNode, resolveDynamicComponent, createTextVNode, toDisplayString, createVNode, createSlots, normalizeProps, guardReactiveProps, normalizeStyle } from 'vue';
import { ElMessage, ElForm, ElCard, ElIcon, ElButton } from 'element-plus';
import '../../../hooks/index.mjs';
import { filterSlots, getLabelSlotName, getFieldSlotName, getExtraSlotName } from '../../utils/index.mjs';
import PlusFormContent from './form-content.vue.mjs';
import { useLocale } from '../../../hooks/useLocale.mjs';

const _hoisted_1 = { class: "plus-form__group__item__icon" };
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "PlusForm"
  },
  __name: "index",
  props: {
    modelValue: { type: null, required: false, default: () => ({}) },
    defaultValues: { type: null, required: false, default: () => ({}) },
    columns: { type: Array, required: false, default: () => [] },
    labelWidth: { type: String, required: false, default: "80px" },
    labelPosition: { type: String, required: false, default: "left" },
    rowProps: { type: Object, required: false, default: () => ({}) },
    colProps: { type: Object, required: false, default: () => ({}) },
    labelSuffix: { type: String, required: false, default: ":" },
    hasErrorTip: { type: Boolean, required: false, default: true },
    hasFooter: { type: Boolean, required: false, default: true },
    hasReset: { type: Boolean, required: false, default: true },
    hasLabel: { type: Boolean, required: false, default: true },
    submitText: { type: String, required: false, default: "" },
    resetText: { type: String, required: false, default: "" },
    submitLoading: { type: Boolean, required: false, default: false },
    footerAlign: { type: String, required: false, default: "left" },
    rules: { type: null, required: false, default: () => ({}) },
    group: { type: [Boolean, Array], required: false, default: false }
  },
  emits: ["update:modelValue", "submit", "change", "reset", "submitError"],
  setup(__props, { expose: __expose, emit }) {
    const props = __props;
    const { t } = useLocale();
    const formInstance = ref();
    const state = reactive({
      values: { ...props.modelValue },
      subColumns: []
    });
    const filterHide = (columns) => {
      return columns.filter((item) => unref(item.hideInForm) !== true);
    };
    const model = computed(() => state.values);
    const style = computed(() => ({
      justifyContent: props.footerAlign === "left" ? "flex-start" : props.footerAlign === "center" ? "center" : "flex-end"
    }));
    state.subColumns = computed(() => filterHide(props.columns));
    const slots = useSlots();
    const labelSlots = filterSlots(slots, getLabelSlotName());
    const fieldSlots = filterSlots(slots, getFieldSlotName());
    const extraSlots = filterSlots(slots, getExtraSlotName());
    watch(
      () => props.modelValue,
      (val) => {
        state.values = val;
      },
      {
        immediate: true
      }
    );
    const handleChange = (_, column) => {
      emit("change", state.values, column);
      emit("update:modelValue", state.values);
    };
    const clearValidate = () => {
      var _a;
      (_a = formInstance.value) == null ? void 0 : _a.clearValidate();
    };
    const handleSubmit = async () => {
      var _a, _b, _c;
      try {
        const valid = await ((_a = formInstance.value) == null ? void 0 : _a.validate());
        if (valid) {
          emit("submit", state.values);
          return true;
        }
      } catch (errors) {
        if (props.hasErrorTip) {
          ElMessage.closeAll();
          const values = Object.values(errors);
          ElMessage.warning(((_c = (_b = values[0]) == null ? void 0 : _b[0]) == null ? void 0 : _c.message) || t("plus.form.errorTip"));
        }
        emit("submitError", errors);
      }
      return false;
    };
    const handleReset = () => {
      clearValidate();
      state.values = { ...props.defaultValues };
      emit("update:modelValue", state.values);
      emit("reset", state.values);
    };
    __expose({
      formInstance,
      handleSubmit,
      handleReset
    });
    return (_ctx, _cache) => {
      return openBlock(), createBlock(unref(ElForm), mergeProps({
        ref_key: "formInstance",
        ref: formInstance,
        rules: _ctx.rules,
        "label-width": _ctx.hasLabel ? _ctx.labelWidth : 0,
        class: ["plus-form", _ctx.hasLabel ? "" : "no-has-label"],
        "label-position": _ctx.labelPosition,
        "validate-on-rule-change": false,
        "label-suffix": _ctx.hasLabel ? _ctx.labelSuffix : ""
      }, _ctx.$attrs, { model: model.value }), {
        default: withCtx(() => [
          renderSlot(_ctx.$slots, "default", {}, () => [
            createCommentVNode(" \u5206\u7EC4\u8868\u5355 "),
            _ctx.group ? (openBlock(true), createElementBlock(
              Fragment,
              { key: 0 },
              renderList(_ctx.group, (groupItem) => {
                return openBlock(), createBlock(
                  unref(ElCard),
                  {
                    key: groupItem.title,
                    class: "plus-form__group__item"
                  },
                  {
                    header: withCtx(() => [
                      renderSlot(_ctx.$slots, "group-header", {
                        title: groupItem.title,
                        columns: groupItem.columns,
                        icon: groupItem.icon
                      }, () => [
                        createElementVNode("div", _hoisted_1, [
                          groupItem.icon ? (openBlock(), createBlock(
                            unref(ElIcon),
                            { key: 0 },
                            {
                              default: withCtx(() => [
                                (openBlock(), createBlock(resolveDynamicComponent(groupItem.icon)))
                              ]),
                              _: 2
                              /* DYNAMIC */
                            },
                            1024
                            /* DYNAMIC_SLOTS */
                          )) : createCommentVNode("v-if", true),
                          createTextVNode(
                            " " + toDisplayString(groupItem.title),
                            1
                            /* TEXT */
                          )
                        ])
                      ])
                    ]),
                    default: withCtx(() => [
                      createVNode(PlusFormContent, {
                        modelValue: state.values,
                        "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => state.values = $event),
                        "row-props": _ctx.rowProps,
                        "col-props": _ctx.colProps,
                        columns: filterHide(groupItem.columns),
                        onChange: handleChange
                      }, createSlots({
                        _: 2
                        /* DYNAMIC */
                      }, [
                        renderList(unref(labelSlots), (_, key) => {
                          return {
                            name: key,
                            fn: withCtx((data) => [
                              renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                            ])
                          };
                        }),
                        renderList(unref(fieldSlots), (_, key) => {
                          return {
                            name: key,
                            fn: withCtx((data) => [
                              renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                            ])
                          };
                        }),
                        renderList(unref(extraSlots), (_, key) => {
                          return {
                            name: key,
                            fn: withCtx((data) => [
                              renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                            ])
                          };
                        })
                      ]), 1032, ["modelValue", "row-props", "col-props", "columns"])
                    ]),
                    _: 2
                    /* DYNAMIC */
                  },
                  1024
                  /* DYNAMIC_SLOTS */
                );
              }),
              128
              /* KEYED_FRAGMENT */
            )) : (openBlock(), createElementBlock(
              Fragment,
              { key: 1 },
              [
                createCommentVNode(" \u666E\u901A\u8868\u5355 "),
                createVNode(PlusFormContent, {
                  modelValue: state.values,
                  "onUpdate:modelValue": _cache[1] || (_cache[1] = ($event) => state.values = $event),
                  "row-props": _ctx.rowProps,
                  "col-props": _ctx.colProps,
                  columns: state.subColumns,
                  "has-label": _ctx.hasLabel,
                  onChange: handleChange
                }, createSlots({
                  _: 2
                  /* DYNAMIC */
                }, [
                  renderList(unref(labelSlots), (_, key) => {
                    return {
                      name: key,
                      fn: withCtx((data) => [
                        renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                      ])
                    };
                  }),
                  renderList(unref(fieldSlots), (_, key) => {
                    return {
                      name: key,
                      fn: withCtx((data) => [
                        renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                      ])
                    };
                  }),
                  renderList(unref(extraSlots), (_, key) => {
                    return {
                      name: key,
                      fn: withCtx((data) => [
                        renderSlot(_ctx.$slots, key, normalizeProps(guardReactiveProps(data)))
                      ])
                    };
                  }),
                  _ctx.$slots["search-footer"] ? {
                    name: "search-footer",
                    fn: withCtx(() => [
                      renderSlot(_ctx.$slots, "search-footer")
                    ]),
                    key: "0"
                  } : void 0
                ]), 1032, ["modelValue", "row-props", "col-props", "columns", "has-label"])
              ],
              64
              /* STABLE_FRAGMENT */
            ))
          ]),
          _ctx.hasFooter ? (openBlock(), createElementBlock(
            "div",
            {
              key: 0,
              class: "plus-form__footer",
              style: normalizeStyle(style.value)
            },
            [
              renderSlot(_ctx.$slots, "footer", normalizeProps(guardReactiveProps({ handleReset, handleSubmit })), () => [
                _ctx.hasReset ? (openBlock(), createBlock(unref(ElButton), {
                  key: 0,
                  onClick: handleReset
                }, {
                  default: withCtx(() => [
                    createCommentVNode(" \u91CD\u7F6E "),
                    createTextVNode(
                      " " + toDisplayString(_ctx.resetText || unref(t)("plus.form.resetText")),
                      1
                      /* TEXT */
                    )
                  ]),
                  _: 1
                  /* STABLE */
                })) : createCommentVNode("v-if", true),
                createVNode(unref(ElButton), {
                  type: "primary",
                  loading: _ctx.submitLoading,
                  onClick: handleSubmit
                }, {
                  default: withCtx(() => [
                    createCommentVNode(" \u63D0\u4EA4 "),
                    createTextVNode(
                      " " + toDisplayString(_ctx.submitText || unref(t)("plus.form.submitText")),
                      1
                      /* TEXT */
                    )
                  ]),
                  _: 1
                  /* STABLE */
                }, 8, ["loading"])
              ])
            ],
            4
            /* STYLE */
          )) : createCommentVNode("v-if", true)
        ]),
        _: 3
        /* FORWARDED */
      }, 16, ["rules", "label-width", "class", "label-position", "label-suffix", "model"]);
    };
  }
});

export { _sfc_main as default };
