// plus-pro-components/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defaults = require('./defaults.js');
require('./components/index.js');
require('./hooks/index.js');
require('./types/index.js');
require('./constants/index.js');
var index = require('./components/dialog/index.js');
var index$1 = require('./components/pagination/index.js');
var index$2 = require('./components/table/index.js');
var index$3 = require('./components/radio/index.js');
var index$4 = require('./components/date-picker/index.js');
var index$5 = require('./components/descriptions/index.js');
var index$6 = require('./components/display-item/index.js');
var index$7 = require('./components/form/index.js');
var index$8 = require('./components/form-item/index.js');
var index$9 = require('./components/search/index.js');
var index$a = require('./components/dialog-form/index.js');
var index$b = require('./components/drawer-form/index.js');
var index$c = require('./components/page/index.js');
var index$d = require('./components/popover/index.js');
var index$e = require('./components/steps-form/index.js');
var index$f = require('./components/input-tag/index.js');
var index$g = require('./components/breadcrumb/index.js');
var index$h = require('./components/sidebar/index.js');
var index$i = require('./components/header/index.js');
var index$j = require('./components/layout/index.js');
var index$k = require('./components/check-card/index.js');
var index$l = require('./components/check-card-group/index.js');
var useGetOptions = require('./hooks/useGetOptions.js');
var useLocale = require('./hooks/useLocale.js');
var useTable = require('./hooks/useTable.js');
var page = require('./constants/page.js');
var form = require('./constants/form.js');

const install = defaults.default.install;

exports.default = defaults.default;
exports.PlusDialog = index.PlusDialog;
exports.PlusPagination = index$1.PlusPagination;
exports.PlusTable = index$2.PlusTable;
exports.PlusRadio = index$3.PlusRadio;
exports.PlusDatePicker = index$4.PlusDatePicker;
exports.PlusDescriptions = index$5.PlusDescriptions;
exports.PlusDisplayItem = index$6.PlusDisplayItem;
exports.PlusForm = index$7.PlusForm;
exports.PlusFormItem = index$8.PlusFormItem;
exports.PlusSearch = index$9.PlusSearch;
exports.PlusDialogForm = index$a.PlusDialogForm;
exports.PlusDrawerForm = index$b.PlusDrawerForm;
exports.PlusPage = index$c.PlusPage;
exports.PlusPopover = index$d.PlusPopover;
exports.PlusStepsForm = index$e.PlusStepsForm;
exports.PlusInputTag = index$f.PlusInputTag;
exports.PlusBreadcrumb = index$g.PlusBreadcrumb;
exports.PlusSidebar = index$h.PlusSidebar;
exports.PlusSidebarItem = index$h.PlusSidebarItem;
exports.PlusHeader = index$i.PlusHeader;
exports.PlusLayout = index$j.PlusLayout;
exports.PlusCheckCard = index$k.PlusCheckCard;
exports.PlusCheckCardGroup = index$l.PlusCheckCardGroup;
exports.useGetOptions = useGetOptions.useGetOptions;
exports.buildLocaleContext = useLocale.buildLocaleContext;
exports.buildTranslator = useLocale.buildTranslator;
exports.translate = useLocale.translate;
exports.useLocale = useLocale.useLocale;
exports.useTable = useTable.useTable;
exports.DefaultPageInfo = page.DefaultPageInfo;
exports.DefaultPageSizeList = page.DefaultPageSizeList;
exports.TableFormRefInjectionKey = form.TableFormRefInjectionKey;
exports.install = install;
