// plus-pro-components/lib/components/radio/src/index.vue2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var elementPlus = require('element-plus');

var _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    name: "PlusRadio"
  },
  __name: "index",
  props: {
    modelValue: { type: [String, Number, Boolean], required: false, default: "" },
    options: { type: Array, required: true, default: () => [] },
    isCancel: { type: Boolean, required: false, default: true }
  },
  emits: ["change", "update:modelValue"],
  setup(__props, { expose: __expose, emit }) {
    const props = __props;
    const radioInstance = vue.ref();
    const radioGroupInstance = vue.ref();
    const state = vue.reactive({ radio: "" });
    vue.watch(
      () => props.modelValue,
      (val) => {
        state.radio = val;
      },
      { immediate: true }
    );
    const attrs = vue.useAttrs();
    const radioClick = (e, val, fieldItemProps) => {
      if (Reflect.has(attrs, "disabled") || (fieldItemProps == null ? void 0 : fieldItemProps.disabled)) {
        return;
      }
      if (!props.isCancel) {
        return;
      } else {
        e.preventDefault();
      }
      state.radio === val ? state.radio = "" : state.radio = val;
      emit("update:modelValue", state.radio);
      emit("change", state.radio);
    };
    const change = (val) => {
      if (props.isCancel)
        return;
      emit("update:modelValue", val);
      emit("change", val);
    };
    __expose({
      radioInstance,
      radioGroupInstance
    });
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createBlock(vue.unref(elementPlus.ElRadioGroup), vue.mergeProps({
        ref_key: "radioGroupInstance",
        ref: radioGroupInstance,
        modelValue: state.radio,
        "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => state.radio = $event)
      }, _ctx.$attrs), {
        default: vue.withCtx(() => [
          (vue.openBlock(true), vue.createElementBlock(
            vue.Fragment,
            null,
            vue.renderList(_ctx.options, (item) => {
              return vue.openBlock(), vue.createBlock(vue.unref(elementPlus.ElRadio), vue.mergeProps({
                key: item.value,
                ref_for: true,
                ref_key: "radioInstance",
                ref: radioInstance,
                label: item.value
              }, item.fieldItemProps, {
                onClick: ($event) => radioClick($event, item.value, item.fieldItemProps),
                onChange: ($event) => change(item.value)
              }), {
                default: vue.withCtx(() => [
                  vue.createTextVNode(
                    vue.toDisplayString(item.label),
                    1
                    /* TEXT */
                  )
                ]),
                _: 2
                /* DYNAMIC */
              }, 1040, ["label", "onClick", "onChange"]);
            }),
            128
            /* KEYED_FRAGMENT */
          ))
        ]),
        _: 1
        /* STABLE */
      }, 16, ["modelValue"]);
    };
  }
});

exports.default = _sfc_main;
