// plus-pro-components/lib/components/drawer-form/src/index.vue2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var index$1 = require('../../form/index.js');
var elementPlus = require('element-plus');
require('../../../hooks/index.js');
var index = require('../../utils/index.js');
var useLocale = require('../../../hooks/useLocale.js');

var _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    name: "PlusDrawerForm"
  },
  __name: "index",
  props: {
    modelValue: { type: null, required: false, default: () => ({}) },
    visible: { type: Boolean, required: false, default: false },
    drawer: { type: null, required: false, default: () => ({}) },
    size: { type: [String, Number], required: false, default: "540px" },
    form: { type: null, required: false, default: () => ({}) }
  },
  emits: ["update:modelValue", "update:visible", "submit", "change", "cancel", "reset"],
  setup(__props, { expose: __expose, emit }) {
    const props = __props;
    const { t } = useLocale.useLocale();
    const formInstance = vue.ref();
    const computedFormInstance = vue.computed(() => {
      var _a;
      return (_a = formInstance.value) == null ? void 0 : _a.formInstance;
    });
    const drawerInstance = vue.ref();
    const state = vue.ref({});
    const subVisible = vue.ref(false);
    const slots = vue.useSlots();
    const labelSlots = index.filterSlots(slots, index.getLabelSlotName());
    const fieldSlots = index.filterSlots(slots, index.getFieldSlotName());
    const extraSlots = index.filterSlots(slots, index.getExtraSlotName());
    vue.watch(
      () => props.visible,
      (val) => {
        subVisible.value = val;
      },
      {
        immediate: true
      }
    );
    vue.watch(
      () => props.modelValue,
      (val) => {
        state.value = val;
      },
      {
        immediate: true
      }
    );
    const handleChange = (values, column) => {
      emit("update:modelValue", values);
      emit("change", values, column);
    };
    const handleSubmitForm = () => {
      emit("submit", state.value);
    };
    const handleCancel = () => {
      subVisible.value = false;
      emit("update:visible", subVisible.value);
      emit("cancel");
    };
    const handleReset = () => {
      emit("reset");
    };
    __expose({
      drawerInstance,
      formInstance: computedFormInstance
    });
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createBlock(vue.unref(elementPlus.ElDrawer), vue.mergeProps({
        ref_key: "drawerInstance",
        ref: drawerInstance,
        modelValue: subVisible.value,
        "onUpdate:modelValue": _cache[1] || (_cache[1] = ($event) => subVisible.value = $event),
        size: _ctx.size || "540px",
        title: vue.unref(t)("plus.drawerForm.title"),
        "close-on-click-modal": false,
        "close-on-press-escape": false
      }, _ctx.drawer, { onClose: handleCancel }), vue.createSlots({
        default: vue.withCtx(() => [
          vue.createVNode(vue.unref(index$1.PlusForm), vue.mergeProps({
            ref_key: "formInstance",
            ref: formInstance,
            modelValue: state.value,
            "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => state.value = $event),
            "footer-align": "right"
          }, _ctx.form, {
            onSubmit: handleSubmitForm,
            onChange: handleChange,
            onReset: handleReset
          }), vue.createSlots({
            _: 2
            /* DYNAMIC */
          }, [
            _ctx.$slots["form-footer"] ? {
              name: "footer",
              fn: vue.withCtx(() => [
                vue.renderSlot(_ctx.$slots, "form-footer")
              ]),
              key: "0"
            } : void 0,
            _ctx.$slots["form-group-header"] ? {
              name: "group-header",
              fn: vue.withCtx(() => [
                vue.renderSlot(_ctx.$slots, "form-group-header")
              ]),
              key: "1"
            } : void 0,
            vue.renderList(vue.unref(labelSlots), (_, key) => {
              return {
                name: key,
                fn: vue.withCtx((data) => [
                  vue.renderSlot(_ctx.$slots, key, vue.normalizeProps(vue.guardReactiveProps(data)))
                ])
              };
            }),
            vue.renderList(vue.unref(fieldSlots), (_, key) => {
              return {
                name: key,
                fn: vue.withCtx((data) => [
                  vue.renderSlot(_ctx.$slots, key, vue.normalizeProps(vue.guardReactiveProps(data)))
                ])
              };
            }),
            vue.renderList(vue.unref(extraSlots), (_, key) => {
              return {
                name: key,
                fn: vue.withCtx((data) => [
                  vue.renderSlot(_ctx.$slots, key, vue.normalizeProps(vue.guardReactiveProps(data)))
                ])
              };
            })
          ]), 1040, ["modelValue"])
        ]),
        _: 2
        /* DYNAMIC */
      }, [
        _ctx.$slots["drawer-header"] ? {
          name: "header",
          fn: vue.withCtx(() => [
            vue.renderSlot(_ctx.$slots, "drawer-header")
          ]),
          key: "0"
        } : void 0,
        _ctx.$slots["drawer-footer"] ? {
          name: "footer",
          fn: vue.withCtx(() => [
            vue.renderSlot(_ctx.$slots, "drawer-footer")
          ]),
          key: "1"
        } : void 0
      ]), 1040, ["modelValue", "size", "title"]);
    };
  }
});

exports.default = _sfc_main;
